<a href="https://colab.research.google.com/github/AshbeeKim/dadaiksunTeamProject/blob/main/ipynb/AshbeeKim/Proj_WeatherGo(Part_Ashebee).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project Flow

## 팀 디자인
  * 구성 :
    * 팀명 : 다다익선(多多익선)
    * 주제 : 기상 데이터에 기반하여 조회 시점 기준 관광지 추천 시스템 구축
    * 팀장 : yunyoseob(윤요섭)
    * 팀원 :
      * seonhong416(김선홍)
      * qkrwjdduf159(박정열)
      * AshbeeKim(김수빈)
  * 분업 : 
    * 논문 : yunyoseob(윤요섭), seonhong416(김선홍)
      * 사전 논문 조사
      * 특정 지역(해운대, 서울대공원)의 기후와 관광 데이터로 논문 기반 가설 검증
    * 수집 : qkrwjdduf159(박정열), AshbeeKim(김수빈)
      * 기후 및 관광 데이터 수집
      * 수집한 데이터 정확도 검증(결측치, 이상치 여부 확인)
    * 분석 : 전원
      * yunyoseob(윤요섭) : 대전 세종 경남 제주 경북
      * seonhong416(김선홍) : 서울 부산 대구 인천
      * qkrwjdduf159(박정열) : 충북 충남 전북 전남
      * AshbeeKim(김수빈) : 울산 경기 광주 강원
    * 디자인 : 팀 로고(초기안 ; 지인 의뢰, 수정중 ; AshbeeKim(김수빈))

## ~~데이터 생성~~
  * 담당 : 없음 (<b>현 프로젝트의 규모 상 진행 불가</b>)
  * 사유1 : 검증할 데이터 수집은 기상청과 관광공사에서 제공
  * 사유2 : 검증할 기간 및 지역 범위가 좁다면 관광객 수를 수집 후 생성할 수도 있지만, 범위가 넒음

## 데이터 수집
  * 주요 데이터 : 기상 & 관광 데이터
  ### 방법
    * internal data : ETL
    * external data : Crawling(Scraping)+ETL
  ### Crawling(Scraping)
    * Web Scraping
    * OpenApi
  ### ETL
    * 검증1 : 기상 데이터(지상 종관 ASOS) 정확도 99% 수렴<small>__(기상청 주장)__</small>
    * 검증2 : <b><공공기관(기상청)</b>에서 관측 및 수집한 자료의 정확도는 그들의 주장 대비 높은 편이 아니었음<small>[完)OpenApi와 해당 사이트의 csv파일 다운 뒤 교차 확인]</small>

## 데이터 분류
### 기상데이터
  * 수집기관 : <b>KMA(한국기상청)</b>
  * 수집방법 : <b>OpenApi</b><small>[지상(종관, ASOS)일자료 조회서비스]</small>
  * 수집기간 : 1/1/2011 ~ 31/12/2020
  * 관리부서 : 경기>서울 or 전북
  * 지점 : 동두천, 이천, 양평, 수원, 파주>...
  * 기간구분 : 1/1/2011 ~ 31/12/2019<small>(사유 : 코로나 이후 급감한 관광객과 매칭 불가)</small>
### 관광데이터
  * 수집기관 : <b>한국관광공사</b>
  * 수집방법 : <b><li>requests</b><small>[주요관광지점입장객 월 통계자료]</small>
  * 수집기간 : 1/1/2011 ~ 31/12/2019

## 데이터 분석
### EDA(동시 진행 중)
### VSL(동시 진행 중)
### Model(동시 진행 중)


# Ashbee Kim's DA Flow__울산, 광주, 경기, 강원 지역 <small>__ver.etude__</small>
  * 1차 마감일 : 2nd.Jul.2021
  * 2차 마감일 : 5th.Jul.2021

## ETL
  * Extract
    * 가설(세부) 검증에 필요한 데이터(칼럼) 정의
    * 가설 검증에 필요한 데이터 추출
  * Transform
    * 기상 데이터는 년>월>일별로 나오지만, 관광데이터는 년>월별로 나옴
    * 지역별  맞춤화 작업 필수
  * Load
    * 현 시점에선 구글 개인 혹은 공유 폴더에 적재 중

## 데이터 분석
### EDA(동시 진행 중)
### VSL(동시 진행 중)
### Model(보류)
  * 전처리 및 시각화 후 모델 특정 가능

# Access to Google Drive for co-working

* Root

In [ ]:
import os
from google.colab import drive
from os.path import join  

# 구글 드라이브 접근
ROOT = "/content/drive"
drive.mount(ROOT)

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


* Private

In [ ]:
# 개인 폴더
PROJECT = "MyDrive/Task/Proj_WeatherGo"
PROJ_PATH = join(ROOT, PROJECT)

CHK_PATH = join(PROJ_PATH, "MINE")
PV_SV = join(PROJ_PATH, "re-ASOS")

* Co-Work

In [ ]:
# 공유 폴더
CO_WORK = "MyDrive/Proj_WT"
CW_PATH = join(ROOT, CO_WORK)

## 논문 공유 폴더
PAPERS = "Papers"   
PAPER_PATH = join(CW_PATH, PAPERS)
# !mkdir "{PAPER_PATH}"

## 데이터 공유 폴더
DATAS = "DataSets"  
DATA_PATH = join(CW_PATH, DATAS)
# !mkdir "{DATA_PATH}"
### 데이터 중 ASOS에 접근
CW_SV = join(DATA_PATH, "기상청_일별_지상_종관_ASOS")

## 분석 공유 폴더
DA = "Analysis"     
DA_PATH = join(CW_PATH, DA)
# !mkdir "{DA_PATH}"

## 코드/마크다운 공유 및 에러 해결 목적(현 공유 폴더 내 없음)
# CODES = "Codes"     
# CODE_PATH = join(CW_PATH, CODES)
## !mkdir "{CODE_PATH}"

# Baseline

## 현재 진행 상황 확인용

In [ ]:
print("현재 공유 폴더 내에 존재하는 디렉토리 :\n")
%cd "{CW_PATH}"
!ls

print("현재 논문 폴더 내에 존재하는 논문 공유 현황 :\n")
%cd "{PAPER_PATH}"
!ls

print("현재 데이터 폴더 내에 존재하는 데이터 수집 현황 :\n")
%cd "{DATA_PATH}"
!ls

print("\n현재 개인 프로젝트 폴더 내에 존재하는 파일 :\n")
%cd "{PROJ_PATH}" # 내가 할 수 있는 선까지의 분석을 하기 위해 만든 폴더(공유파일에 없음)
!ls

print("\n현재 분석 폴더 내에 존재하는 파일 현황 :\n")
%cd "{DA_PATH}"
!ls

# print("현재 코드 공유 폴더 내에 존재하는 파일 :\n")
# %cd "{CODE_PATH}"
# !ls


현재 공유 폴더 내에 존재하는 디렉토리 :

/content/drive/.shortcut-targets-by-id/1TVtWvZ4DNn7V20Aj6JzSd5_iRdbjEXbG/Proj_WT
 Analysis  '공유폴더 내 데이터 활용해서 모델 분석할 때 편한 방법.ipynb'
 DataSets   Papers
현재 논문 폴더 내에 존재하는 논문 공유 현황 :

/content/drive/.shortcut-targets-by-id/1TVtWvZ4DNn7V20Aj6JzSd5_iRdbjEXbG/Proj_WT/Papers
'해외방문관광객의 목적지 선택에 대한 기후의 영향 (1).pdf'
'빅데이터 기반 관광지 추천 시스템 구현.hwp'
 논문_리뷰.ipynb
'한국형 관광기후지수(KTCI)의 개발에 관한 연구.gdoc'
'한국형 관광기후지수(KTCI)의 개발에 관한 연구.pdf'
 빅데이터기반관광지추천시스템구현한국관광공사LOD를중심으로.pdf
 관광업체관광지추천유형정리.pptx
'날씨 및 요일 특성이 관광정보시스템 이용에 미치는 영향에 관한 연구.rtf'
'기후인자가 관광수요에 미치는 영향 패널데이터 모형의 적용.rtf'
'Weather Go ( 웨더고) 초기 회의록(20210621).hwp'
'Weather Go (웨더고) 두번째 회의록.pptx'
현재 데이터 폴더 내에 존재하는 데이터 수집 현황 :

/content/

## 기본 라이브러리

In [ ]:
!pip install xmltodict

In [ ]:
import sys
import re
import math
import itertools

import random
import time, datetime
from datetime import date, datetime, timedelta

import csv
import numpy as np
import pandas as pd

from IPython.display import display
import matplotlib.pyplot as plt
import seaborn as sns
# 조회한 자료를 ppt파일에 참고할 자료로 구상할 경우, 
  # 빈도수 혹은 총 날씨 관련 특허 몇 건 중 몇 건,
  # 관광 관련 특허 총 몇 건 중 몇 건으로 (%)이미지 생성

# Colab Python에서 R환경 번갈아가면서 쓸 때 사용
import rpy2
# %load_ext rpy2.ipython
# R 환경을 사용하려는 모든 셀에 아래와 같이 선언해야 함
  # (해당 셀이 선언되면 안 되기 때문에 주석처리)
# %%R

import json #json을 다루기 위한 모듈
from collections import OrderedDict
from pprint import pprint #json을 좀 더 보기 편하게 출력

import urllib
import urllib.request
from urllib.request import urlopen
from urllib.parse import urlencode, unquote, quote_plus
import requests
import xmltodict
    ## xml형식으로만 지원하는 api를 csv파일로 저장하려면, parsing한 xml를 dict로 변환해야 함
#import xml.dom.minidom 
    ## 다른 언어와  유사한 API를 갖는 문서 객체 모델 인터페이스의 최소 구현
    ## 악의적으로 구성된 데이터로부터 안전하지 않음
#from xml.dom.minidom import parse, parseString
from xml.etree.ElementTree import parse
    ## DOM에 아직 능숙하지 않은 사용자는 XML처리에 "xml.etree.ElementTree" 모듈을 사용하는 것을 고려

In [ ]:
# # 시각화 시, 나눔 폰트 설치
# !apt-get install fonts-nanum*
# !apt-get install fontconfig
# !fc-cache -fv
# # !cp /usr/share/fonts/truetype/nanum/Nanum* /usr/local/lib/python2.7/dist-packages/matplotlib/mpl-data/fonts/ttf/
# # !cp /usr/share/fonts/truetype/nanum/Nanum* /usr/local/lib/python3.6/dist-packages/matplotlib/mpl-data/fonts/ttf/
# !cp /usr/share/fonts/truetype/nanum/Nanum* /usr/local/lib/python3.7/dist-packages/matplotlib/mpl-data/fonts/ttf/
# !rm -rf /content/.cache/matplotlib/*

# # 그래프 활용
# import matplotlib.pyplot as plt
#   import matplotlib as mpl
#   import matplotlib.font_manager as fm
#   import numpy as np
	
#   %matplotlib inline
#   %config InlineBackend.figure_format = 'retina'
	
#   mpl.rcParams['axes.unicode_minus'] = False
#   # 그래프에서 마이너스 폰트 깨질 경우 대비
	
#   path = '/usr/share/fonts/truetype/nanum/NanumGothicBold.ttf'
#   font_name = fm.FontProperties(fname=path, size=18).get_name()
#   plt.rc('font', family=font_name)
#   fm._rebuild() # 이걸 해줘야 plt.rc가 작동

#   plt.plot(np.random.randn(4, 8), np.random.randn(4,8), 'bo--')
#   plt.title('타이틀')
#   plt.xlabel('X 라벨')
#   plt.ylabel('Y 라벨')
#   plt.show()



### 근데 다른 방법을 사용했을 때가 훨씬 편했던 듯....

In [ ]:
data = pd.read_excel(f'{CW_SV}/branch.xlsx')
data['지점'] = data['지점'].astype('str')
office = data['관리관서'].unique()
print(office)
branch = []
for j in office:
  KMA = {}
  for i in data.index:
    if data.loc[i,'관리관서'] == j:
      KMA[data.loc[i,'지점명']] = data.loc[i,'지점']
  branch.append(KMA)

for of, br in zip(office, branch):
  print(of, br)
  for name, code in br.items():
    print(name, code)

['강원지방기상청' '춘천기상대' '수도권기상청' '대구지방기상청' '청주기상지청' '홍성기상대' '안동기상대' '대전지방기상청'
 '전주기상지청' '울산기상대' '창원기상대' '광주지방기상청' '부산지방기상청' '목포기상대' '제주지방기상청' '인천기상대']
강원지방기상청 {'속초': '90', '철원': '95', '대관령': '100', '북강릉': '104', '강릉': '105', '동해': '106', '원주': '114', '영월': '121', '인제': '211', '태백': '216', '정선군': '217'}
속초 90
철원 95
대관령 100
북강릉 104
강릉 105
동해 106
원주 114
영월 121
인제 211
태백 216
정선군 217
춘천기상대 {'북춘천': '93', '춘천': '101', '홍천': '212'}
북춘천 93
춘천 101
홍천 212
수도권기상청 {'동두천': '98', '파주': '99', '백령도': '102', '서울': '108', '인천': '112', '수원': '119', '양평': '202', '이천': '203'}
동두천 98
파주 99
백령도 102
서울 108
인천 112
수원 119
양평 202
이천 203
대구지방기상청 {'울릉도': '115', '상주': '137', '포항': '138', '대구': '143', '봉화': '271', '청송군': '276', '영덕': '277', '의성': '278', '구미': '279', '영천': '281', '경주시': '283'}
울릉도 115
상주 137
포항 138
대구 143
봉화 271
청송군 276
영덕 277
의성 278
구미 279
영천 281
경주시 283
청주기상지청 {'충주': '127', '청주': '131', '추풍령': '135', '제천': '221', '보은': '226'}
충주 127
청주 131
추풍령 135
제천 221
보은 226
홍성기상대 {'서산': '129', '홍성': '177'}
서산 129
홍성 

# 데이터 수집

## ASOS(Open Api)

In [ ]:
# SERVICE = "AsosDalyInfoService"
# STYPE = "getWthrDataList"
# # KEY = "{Encoding Key}"  #서비스키
# KEY = "yRxjHtUyF3aadfyII7o4rlbGO0QK4kY2Uf46AoDzfvoCgVT6wHHceyMx%2Be1DZvcYHLAl4h2ANO02EFyWxkd%2FHA%3D%3D"
# PG_NO = "1"
# NoRows = "10"
# Dtype = ["XML", "JSON"]
# DCd = "ASOS"
# DtCd = "DAY"

# StDt = ["20110101", "20120101", "20130101", "20140101", "20150101", "20160101", "20170101", "20180101", "20190101"]
# EdDt = ["20111231", "20121231", "20131231", "20141231", "20151231", "20161231", "20171231", "20181231", "20191231"]

In [ ]:
# 1년치 데이터 수집 시 에러 발견해서 주말에 수정 필요

# for of, br in zip(office, branch):
#     locals()[f'NoData_{of}'] = []
#     locals()[f'df_{of}'] = []
  
#     for stdt, eddt in zip(StDt, EdDt):
#         print(f'{stdt[:4]}년도\n{of} 관리관서 데이터 수집을 시작합니다.')

#         for name, code in br.items():
#             print(f'\t{stdt[:4]}년도 {name}지점({code}) 데이터 수집을 시작합니다.')
#             try:
#                 url = f"http://apis.data.go.kr/1360000/{SERVICE}/{STYPE}?serviceKey={KEY}&pageNo={PG_NO}&numOfRows={NoRows}&dataType={Dtype[1]}&dataCd={DCd}&dateCd={DtCd}&startDt={stdt}&endDt={eddt}&stnIds={code}"
#                 rep = requests.get(url)
#                 time.sleep(random.randint(2,10))
#                 origin = rep.json()

#                 if origin['response']['body']['totalCount']=="0":
#                     print(f"They don't have Data at {name}({code})")

#                 else:
#                     NoRows = origin['response']['body']['totalCount']
#                     url = f"http://apis.data.go.kr/1360000/{SERVICE}/{STYPE}?serviceKey={KEY}&pageNo={PG_NO}&numOfRows={NoRows}&dataType={Dtype[1]}&dataCd={DCd}&dateCd={DtCd}&startDt={stdt}&endDt={eddt}&stnIds={code}"
#                     rep = requests.get(url)
#                     time.sleep(random.randint(2,10))
#                     origin = rep.json()
#                     locals()[f'df_{of}'] = pd.concat([locals()[f'df_{of}'], pd.json_normalize(origin['response']['body']['items']['item'])])

#                     print(f'\t{eddt[:4]}년도까지 {name}지점({code}) 데이터 수집이 완료되었습니다.')                      

#             except:
#                 print(f'\t{"*"*5}{eddt[:4]}년도 {name}지점({code}) 데이터 수집 중 에러가 발생했습니다.{"*"*5}')
#                 pass

#     print(f'총 {eddt[:4]}년도\n{of} 관리관서 데이터 수집이 완료되었습니다')

## re-ASOS

api를 사용하자니, 년>월>일로 접근할 때 어떤 기준으로 데이터를 처리해야 할 지에 대한 가늠이 안 됨

불신 요인 : 결측일에 대한 불신

In [ ]:
from google.colab import files

%cd "{PV_SV}"
files.upload()

/content/drive/.shortcut-targets-by-id/1srYR8TqwFsUSR4BYzQtIBmWUKrEBjd44/Task/Proj_WeatherGo/re-ASOS


Saving SUDOGUN_Y_ASOS.csv.csv to SUDOGUN_Y_ASOS.csv.csv
Saving GYTBSHI_Y_ASOS.csv.csv to GYTBSHI_Y_ASOS.csv.csv
Saving GYTBSHI_YM_ASOS.csv.csv to GYTBSHI_YM_ASOS.csv.csv
Saving SUDOGUN_YM_ASOS.csv.csv to SUDOGUN_YM_ASOS.csv.csv


{'GYTBSHI_YM_ASOS.csv.csv': b'\xc1\xf6\xc1\xa1,\xc1\xf6\xc1\xa1\xb8\xed,\xc0\xcf\xbd\xc3,\xc6\xf2\xb1\xd5\xb1\xe2\xbf\xc2(\xa1\xc6C),\xc6\xf2\xb1\xd5\xc3\xd6\xb0\xed\xb1\xe2\xbf\xc2(\xa1\xc6C),\xc6\xf2\xb1\xd5\xc3\xd6\xc0\xfa\xb1\xe2\xbf\xc2(\xa1\xc6C),\xc3\xd6\xb0\xed\xb1\xe2\xbf\xc2(\xa1\xc6C),\xc3\xd6\xc0\xfa\xb1\xe2\xbf\xc2(\xa1\xc6C),\xc3\xd6\xb0\xed\xb1\xe2\xbf\xc2 \xb3\xaa\xc5\xb8\xb3\xad\xb3\xaf(yyyymmdd),\xc3\xd6\xc0\xfa\xb1\xe2\xbf\xc2 \xb3\xaa\xc5\xb8\xb3\xad\xb3\xaf(yyyymmdd),\xc6\xf2\xb1\xd5\xc7\xf6\xc1\xf6\xb1\xe2\xbe\xd0(hPa),\xc6\xf2\xb1\xd5\xc7\xd8\xb8\xe9\xb1\xe2\xbe\xd0(hPa),\xc3\xd6\xb0\xed\xc7\xd8\xb8\xe9\xb1\xe2\xbe\xd0(hPa),\xc3\xd6\xc0\xfa\xc7\xd8\xb8\xe9\xb1\xe2\xbe\xd0(hPa),\xc3\xd6\xb0\xed\xc7\xd8\xb8\xe9\xb1\xe2\xbe\xd0 \xb3\xaa\xc5\xb8\xb3\xad\xb3\xaf(yyyymmdd),\xc3\xd6\xc0\xfa\xc7\xd8\xb8\xe9\xb1\xe2\xbe\xd0 \xb3\xaa\xc5\xb8\xb3\xad\xb3\xaf(yyyymmdd),\xc6\xf2\xb1\xd5\xbc\xf6\xc1\xf5\xb1\xe2\xbe\xd0(hPa),\xc3\xd6\xb0\xed\xbc\xf6\xc1\xf5\xb1\xe2\xbe\xd0(hPa)

In [ ]:
  !ls

GangwonYearASOS.csv  GuangjuYMASOS.csv	   GyeonggiYMDASOS.csv
GangwonYMASOS.csv    GuangjuYMDASOS.csv    UlsanYearASOS.csv
GangwonYMDASOS.csv   GyeonggiYearASOS.csv  UlsanYMASOS.csv
GuangjuYearASOS.csv  GyeonggiYMASOS.csv    UlsanYMDASOS.csv


In [ ]:
drangelst = ["Ulsan", "Guangju", "Gangwon", "Gyeonggi"]
dtimen_lst = ["Year", "YM", "YMD"]
SV = "ASOS"

wt = "WT"

for area, dt_type in itertools.product(darea_lst, dtimen_lst):
  path = PV_SV+"/"+area+dt_type+SV
  locals()[f"{area}_WT_{dt_type}"] = pd.read_csv(f"{path}.csv", encoding="cp949")

FileNotFoundError: ignored

In [ ]:
%cd "{PV_SV}"
!ls

/content/drive/.shortcut-targets-by-id/1srYR8TqwFsUSR4BYzQtIBmWUKrEBjd44/Task/Proj_WeatherGo/re-ASOS
GangwonYearASOS.csv  GyeonggiYearASOS.csv  SUDOGUN_YM_ASOS.csv
GangwonYMASOS.csv    GyeonggiYMASOS.csv    UlsanYearASOS.csv
GangwonYMDASOS.csv   GyeonggiYMDASOS.csv   UlsanYMASOS.csv
GuangjuYearASOS.csv  GYTBSHI_Y_ASOS.csv    UlsanYMDASOS.csv
GuangjuYMASOS.csv    GYTBSHI_YM_ASOS.csv
GuangjuYMDASOS.csv   SUDOGUN_Y_ASOS.csv


In [ ]:
# !mv SUDOGUN_YM_ASOS.csv.csv SUDOGUN_YM_ASOS.csv
# !mv GYTBSHI_Y_ASOS.csv.csv GYTBSHI_Y_ASOS.csv
# !mv GYTBSHI_YM_ASOS.csv.csv GYTBSHI_YM_ASOS.csv
# !mv SUDOGUN_Y_ASOS.csv.csv SUDOGUN_Y_ASOS.csv

In [ ]:
darea_lst = ["SUDOGUN", "GYTBSHI"]
dtime_lst = ["Y", "YM"]
SV = "ASOS"

wt = "WT"

for area, dt_type in itertools.product(darea_lst, dtime_lst):
  path = PV_SV+"/"+area+"_"+dt_type+"_"+SV
  locals()[f"{area}_WT_{dt_type}"] = pd.read_csv(f"{path}.csv", encoding="cp949")

## tour.go.kr

# 데이터 처리

* 기관 내 변동사항에 따른 결측치 발생 유무
* 결측치와 이상치 처리 근거

## 행정정보(dict)

### 기상 행정

In [ ]:
data = pd.read_excel(f'{CW_SV}/branch.xlsx')
data['지점'] = data['지점'].astype('str')
office = data['관리관서'].unique()
print(office)
branch = [] # branch : 기상청 지점
for j in office: #office : 기상청 관리관서
  KMA = {}
  for i in data.index:
    if data.loc[i,'관리관서'] == j:
      KMA[data.loc[i,'지점명']] = data.loc[i,'지점']
  branch.append(KMA)

['강원지방기상청' '춘천기상대' '수도권기상청' '대구지방기상청' '청주기상지청' '홍성기상대' '안동기상대' '대전지방기상청'
 '전주기상지청' '울산기상대' '창원기상대' '광주지방기상청' '부산지방기상청' '목포기상대' '제주지방기상청' '인천기상대']


### 관광 행정

In [ ]:
#서울
Seoul_ID = {"서울 강남구":"1168000000", "서울 강동구":"1174000000", "서울 강북구":"1130500000", "서울 강서구":"1150000000",
            "서울 관악구":"1162000000", "서울 광진구":"1121500000", "서울 구로구":"1153000000", "서울 금천구":"1154500000",
            "서울 노원구":"1135000000", "서울 도봉구":"1132000000", "서울 동대문구":"1123000000", "서울 동작구":"1159000000",
            "서울 마포구":"1144000000", "서울 서대문구":"1141000000", "서울 서초구":"1165000000", "서울 성동구":"1120000000",
            "서울 성북구":"1129000000", "서울 송파구":"1171000000", "서울 양천구":"1147000000", "서울 영등포구":"1156000000", 
            "서울 용산구":"1117000000", "서울 은평구":"1138000000", "서울 종로구":"1111000000", "서울 중구":"1114000000",
            "서울 중랑구":"1126000000"}
#부산
BuSan_ID = {"부산 강서구":"2644000000", "부산 금정구":"2641000000", "부산 기장군":"2671000000",
            "부산 남구":"2629000000", "부산 동구":"2617000000", "부산 동래구":"2626000000",
            "부산 부산진구":"2623000000", "부산 북구":"2632000000", "부산 사상구":"2653000000",
            "부산 사하구":"2638000000", "부산 서구":"2614000000", "부산 수영구":"2650000000",
            "부산 연제구":"2647000000", "부산 영도구":"2620000000", "부산 중구":"2611000000",
            "부산 해운대구":"2635000000"}
#인천
InCheon_ID = {"인천 강화군":"2871000000","인천 계양구":"2824500000", "인천 남구":"2817000000",
              "인천 남동구":"2820000000", "인천 동구":"2814000000", "인천 부평구":"2823700000",
              "인천 서구":"2826000000", "인천 연수구":"2818500000", "인천 옹진구":"2872000000",
              "인천 중구":"2811000000"}
#대구
DaeGu_ID = {"대구 서구":"2717000000", "대구 수성구":"2726000000", "대구 중구":"2711000000",
            "대구 남구":"2720000000", "대구 달서구":"2729000000", "대구 동구":"2714000000",
            "대구 북구":"2723000000", "대구 달성군":"2771000000"}        
#대전
DaeJeon_ID = {"대전 대덕구":"3023000000", "대전 동구":"3011000000", "대전 서구":"3017000000",
              "대전 유성구":"3020000000", "대전 중구":"3014000000"}
#울산
UlSan_ID = {"울산 남구":"3114000000", "울산 동구":"3117000000", "울산 북구":"3020000000",
            "울산 울주군":"3171000000", "울산 중구":"3111000000"}
#광주
GuangJu_ID = {"광구 광산구":"2920000000", "광주 남구":"2915500000", "광주 동구":"2911000000", 
              "광주 북구":"2917000000", "광주 서구":"2914000000"}
#세종시
SeJong = {"세종 세종시":"3611031000"}

#제주도
JeJu = {"제주 서귀포시":"5013000000", "제주 제주시":"5011000000"}

#경기
GyeongGi_ID = {"경기 가평군":"4182000000", "경기 고양시":"4128100000", "경기 과천시":"4129000000", "경기 광명시":"4121000000",
               "경기 광주시":"4161000000", "경기 구리시":"4131000000", "경기 군포시":"4141000000", "경기 김포시":"4157000000",
               "경기 남양주시":"4136000000", "경기 동두천시":"4125000000", "경기 부천시":"4119500000", "경기 성남시":"4113100000",
               "경기 수원시":"4111100000", "경기 시흥시":"4139000000", "경기 안산시":"4127100000", "경기 안성시":"4155000000",
               "경기 안양시":"4117100000", "경기 양주시":"4163000000", "경기 양평군":"4183000000", "경기 여주시":"4167000000",
               "경기 연천군":"4180000000", "경기 오산시":"4137000000", "경기 용인시":"4146100000", "경기 의왕시":"4143000000",
               "경기 의정부시":"4115000000", "경기 이천시":"4150000000", "경기 파주시":"4148000000","경기 평택시":"4122000000",
               "경기 포천시":"4165000000", "경기 하남시":"4145000000", "경기 화성시":"4159000000"}

#강원
GangWon_ID = {"강원 강릉시":"4215000000", "강원 고성군":"4282000000", "강원 동해시":"4217000000", "강원 삼척시":"4223000000",
              "강원 속초시":"4221000000", "강원 양구군":"4280000000", "강원 양양군":"4283000000", "강원 영월군":"4275000000",
              "강원 원주시":"4213000000", "강원 인제군":"4281000000", "강원 정선군":"4277000000", "강원 철원군":"4278000000",
              "강원 춘천시":"4211000000", "강원 태백시":"4219000000", "강원 평창군":"4276000000", "강원 홍천군":"4272000000",
              "강원 화천군":"4279000000", "강원 횡성군":"4274000000"}

#충북
ChungBuk_ID = {"충북 괴산군":"4376000000", "충북 단양군":"4380000000", "충북 보은군":"4372000000", "충북 영동군":"4374000000",
               "충북 옥천군":"4373000000", "충북 음성군":"4377000000", "충북 제천시":"4315000000", "충북 증평군":"4374500000",
               "충북 진천군":"4375000000", "충북 청주시":"4311100000", "충북 충주시":"4313000000", "충북 청원군":"4311400000"}

#충남
ChungNam_ID = {"충남 계룡시":"4425000000", "충남 공주시":"4415000000", "충남 금산군":"4471000000", "충남 논산시":"4423000000",
               "충남 당진시":"4427000000", "충남 보령시":"4418000000", "충남 부여군":"4476000000", "충남 서산시":"4421000000",
               "충남 서천군":"4477000000", "충남 아산시":"4420000000", "충남 예산군":"4481000000", "충남 천안시":"4413100000",
               "충남 청양군":"4479000000", "충남 태안군":"4482500000", "충남 홍성군":"4480000000"}

#전북
JeonBuk_ID = {"전북 고창군":"4579000000", "전북 군산시":"4513000000", "전북 김제시":"4521000000", "전북 남원시":"4519000000",
              "전북 무주군":"4573000000", "전북 부안군":"4580000000", "전북 순창군":"4577000000", "전북 완주군":"4571000000",
              "전북 익산시":"4514000000", "전북 임실군":"4575000000", "전북 장수군":"4574000000", "전북 전주시":"4511100000",
              "전북 정읍시":"4518000000", "전북 진안군":"4572000000"}

#전남
JeonNam_ID = {"전남 강진군":"4681000000", "전남 고흥군":"4677000000", "전남 곡성군":"4672000000", "전남 광양시":"4623000000",
              "전남 구례군":"4673000000", "전남 나주시":"4617000000", "전남 담양군":"4671000000", "전남 목포시":"4611000000",
              "전남 무안군":"4684000000", "전남 보성군":"4678000000", "전남 순천시":"4615000000", "전남 신안군":"4691000000",
              "전남 여수시":"4613000000", "전남 영광군":"4687000000", "전남 영암군":"4683000000", "전남 완도군":"4689000000",
              "전남 장성군":"4688000000", "전남 장흥군":"4680000000", "전남 진도군":"4690000000", "전남 함평군":"4686000000",
              "전남 해남군":"4682000000", "전남 화순군":"4679000000"}

#경북
GyeongBuk_ID = {"경북 경산시":"4729000000", "경북 경주시":"4713000000", "경북 고령군":"4783000000", "경국 구미시":"4719000000",
                "경북 군위군":"4772000000", "경북 김천시":"4715000000", "경북 문경시":"4728000000", "경북 봉화군":"4792000000",
                "경북 상주시":"4725000000", "경북 성주군":"4784000000", "경북 안동시":"4717000000", "경북 영덕군":"4777000000",
                "경북 영양군":"4776000000", "경북 영주시":"4721000000", "경북 영천시":"4723000000", "경북 예천군":"4790000000",
                "경북 울릉군":"4794000000", "경북 울진군":"4793000000", "경북 의성군":"4773000000", "경북 청도군":"4782000000",
                "경북 청송군":"4775000000", "경북 칠곡군":"4785000000", "경북 포항시":"4711100000"}

#경남
GyeongNam_ID = {"경남 거제시":"4831000000", "경남 거창군":"4888000000", "경남 고성군":"4882000000", "경남 김해시":"4825000000",
                "경남 남해군":"4884000000", "경남 밀양시":"4827000000", "경남 사천시":"4824000000", "경남 산청군":"4886000000",
                "경남 양산시":"4833000000", "경남 의령군":"4872000000", "경남 진주시":"4817000000", "경남 창녕군":"4874000000",
                "경남 창원시":"4812100000", "경남 통영시":"4822000000", "경남 하동군":"4885000000", "경남 함안군":"4873000000",
                "경남 함양군":"4887000000", "경남 합천군":"4889000000"}

### 담당 지역

내가 해야할 지역 : 울산, 광주, 강원, 경기

In [ ]:
#관광행정(행정시도군)
# AREA_ID = [UlSan_ID, GuangJu_ID, GangWon_ID, GyeongGi_ID]
#기상행정(지점)


## 결측치 확인 및 처리 함수

In [ ]:
def init_Date():
  print(f"{'='*10}{'<'*5}\tInitializing\t{'>'*5}{'='*300}")
  sY, sM, sD  = map(int, input("수집 시작일 : ").split(" "))
  startDate = datetime(sY, sM, sD)

  eY, eM, eD  = map(int, input("수집 종료일 : ").split(" "))
  endDate = datetime(eY, eM, eD)

  return startDate, endDate

def Y_YM_Duration(startDate, endDate):
  Y_Duration = int(endDate.strftime("%Y"))-int(startDate.strftime("%Y"))+1
  YM_Duration = Y_Duration*12

  return Y_Duration, YM_Duration

def Date_Duration(startDate, endDate):
  # print(f"{'='*10}{'<'*5}\tInitializing\t{'>'*5}{'='*300}")
  # sY, sM, sD  = map(int, input("시작일 : ").split(" "))
  # startDate = datetime(sY, sM, sD)

  # eY, eM, eD  = map(int, input("종료일 : ").split(" "))
  # endDate = datetime(eY, eM, eD)

  duration = (endDate-startDate).days
  print(f"조사 기간의 기상 데이터의 일 수는 총 {duration}일이어야 합니다.\n")
  #결측일이 없으려면, duration 값과 기상 데이터의 rows가 동일해야 함

  return duration#, startDate, endDate


def Missing_Date_Check(dt, datecol, duration, startDate=None, check_type="%Y%m%d"):
  print(f"{'='*10}{'<'*5}\tmissing dates\t{'>'*5}{'='*300}")
  missing_days = []
  missing_day = 0

  for day in range(duration):
    check_date = startDate+timedelta(days=day)
    if dt[datecol][day-missing_day].strftime(check_type)==check_date.strftime(check_type):
      continue
    elif dt[datecol][day-missing_day].strftime(check_type)!=check_date.strftime(check_type):
      missing_day+=1
      missing_days.append(check_date) # 활용을 쉽게 활용할 수 있도록 datetime으로 추가

  if len(missing_days)!=0:
    print(f"*경고*\n결측일이 발견되었습니다.")
    print(f"{[mdate.strftime(check_type) for mdate in missing_days]}의 결측일 여부를 확인하세요.\n")

  return missing_days


def Re_val_MDate(dt, datecol, missing_days, check_type="%Y%m%d"):
  print(f"{'='*10}{'<'*5}\tre-validation\t{'>'*5}{'='*300}")
  cnt = 0
  re_val_date = []

  for idx, mdate in enumerate(missing_days):
    if len(dt[dt[datecol]==mdate.strftime(check_type)].index)!=0:
      re_val_date.append(idx)
    elif len(dt[dt[datecol]==mdate.strftime(check_type)].index)==0:
      cnt += 1

  if cnt==len(missing_days):
    print("재검증 결과 : 상동")
    print(f"해당 기상 데이터에 존재하는 결측일은 총 {cnt}일입니다.\n")
  elif cnt!=len(missing_days):
    for pop_idx in re_val_date:
      missing_days.pop(pop_idx)
    print("재검증 결과 : 상이")
    print(f"해당 기상 데이터에 존재하는 결측일은 총 {len(missing_days)}일입니다.\n")

  return missing_days


def Cnts_MDays_in_YearMonth(missing_days, check_ym_type="%Y%m", return_ym_type="%Y년%m월"):
  print(f"{'='*10}{'<'*5}\tCounts of Missing days by Months\t{'>'*5}{'='*300}")

  if check_ym_type!=return_ym_type:
    missing_days_by_ym = [mdates.strftime(return_ym_type) for mdates in missing_days]
    missing_ym = sorted(list(set(missing_days_by_ym)))
    for mYM in missing_ym:
      globals()[f"missing_{mYM}"] = missing_days_by_ym.count(mYM)
      print(f'{mYM}의 결측일 수 : {globals()[f"missing_{mYM}"]}\n')

  else :
    missing_days_by_ym = [mdates.strftime(check_ym_type) for mdates in missing_days]
    missing_ym = sorted(list(set(missing_days_by_ym)))
    for mYM in missing_ym:
      globals()[f"missing_{mYM}"] = missing_days_by_ym.count(mYM)
      print(f'{mYM}의 결측일 수 : {globals()[f"missing_{mYM}"]}\n')

  return missing_ym

## 중복 변수

In [ ]:
darea_lst = ["Ulsan", "Guangju", "Gangwon", "Gyeonggi"]
dtime_lst = ["Year", "YM", "YMD"]
SV = "ASOS"
wt = "WT"

In [ ]:
stD, edD = init_Date()

==========<<<<<	Initializing	>>>>>============================================================================================================================================================================================================================================================================================================
수집 시작일 : 2011 1 1
수집 종료일 : 2019 12 31


In [ ]:
Yduring, YMduring = Y_YM_Duration(stD, edD) #9, 108

In [ ]:
YMDduring = Date_Duration(stD, edD)

조사 기간의 기상 데이터의 일 수는 총 3286일이어야 합니다.



## ASOS(YMD)

### 울산 기상대>울산

울산기상대->울산

In [ ]:
#울산기상대 csv 데이터 불러오기
YRS = ["1112", "1314", "1516", "1718", "1920"]
name = "울산기상대"
for yrs in YRS:
  locals()[f'df{yrs}_{name[:2]}'] = pd.read_csv(f'{CW_SV}/{yrs}/{name}.csv')

df_울산기상대 = pd.DataFrame()

for yrs in YRS:
  df_울산기상대 = pd.concat([df_울산기상대, locals()[f'df{yrs}_{name[:2]}']])

print(df_울산기상대.shape)

In [ ]:
# 울산 기상청 자료(변수명 : Ulsan_Weather, 기간 : 11~20년)
Ulsan_Weather = df_울산기상대[df_울산기상대.iloc[:, 1]=="울산"]
print(Ulsan_Weather.shape)
Ulsan_Weather.reset_index(drop=True, inplace=True)

Ulsan_Weather["일시"] = Ulsan_Weather["일시"].values.astype("datetime64")

print(display(Ulsan_Weather.head(2)))

(3323, 62)


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """


지점 번호 지점명         일시  평균 기온(°C)  ...  합계 소형증발량(mm)  9-9강수(mm)  일기현상  안개 계속 시간(hr)
0    152  울산 2011-01-01       -1.4  ...           NaN        NaN   NaN           NaN
1    152  울산 2011-01-02        1.8  ...           NaN        NaN   NaN           NaN

[2 rows x 62 columns]

None


일시는 string형태로 되어 있었으나, 시계열에 활용 및 20년도와 구분을 편하게 하기 위해 datetime64로 바꿈

우선 20년도는 분석 기간에 들어가지 않기 때문에, 11~19년과 20년도 데이터로 구분

* 구분
  * 방법 1 : 2020년의 일수를 구한 뒤, iloc[:-일수, :]와 iloc[-일수:, :]로 분류 <small></br>    (예상되는 문제 : 2020년의 모든 요일에 대한 기상 정보가 없을 수 있음)</small>
  * 방법 2 : 2020년 1월 1일 미만의 정보를 쌓은 뒤, 20년은 20년 이전 데이터 shape으로 확인 후 iloc[20년 이전 데이터.shape[0]:, :]으로 생성 <small></br>    (예상되는 문제 : 11년부터 확인 및 쌓는다면, 시간이 오래 걸림;;;)</small>
  * 방법 3 : 2020년 1월 1일 이후의 데이터를 역으로 쌓은 뒤, 20년 데이터 shape으로 확인 뒤 iloc[:20년 이후 데이터.shape[0], :]으로 생성 <small></br>    (현재로는 이 방법이 제일 빠른 것으로 사료됨)</small>

In [ ]:
# 11~19년도와 20년 기상 데이터(변수명 : Ulsan_WT, Ulsan_WT20)
cnt = 0
for idx in Ulsan_Weather[::-1].index:
  if Ulsan_Weather["일시"][idx].strftime("%Y%m%d")>="20200101":
      cnt+=1
  elif Ulsan_Weather["일시"][idx].strftime("%Y%m%d")<"20200101":
    break

print(f"20년도의 정보가 들어가 있는 {cnt}일 동안의 데이터는 따로 분류하겠습니다.")


Ulsan_WT = Ulsan_Weather.iloc[:len(Ulsan_Weather)-cnt, :]
Ulsan_WT20 = Ulsan_Weather.iloc[len(Ulsan_Weather)-cnt:, :]
print(f"\nshape of 20년도 데이터(Ulsan_WT20) : {Ulsan_WT20.shape}") #2020년은 366일이니까 이상치 아님
print(display(Ulsan_WT.tail(2)))


Ulsan_WT20.reset_index(drop=True, inplace=True)

20년도의 정보가 들어가 있는 366일 동안의 데이터는 따로 분류하겠습니다.

shape of 20년도 데이터(Ulsan_WT20) : (366, 62)


지점 번호  ... 안개 계속 시간(hr)
2955    152  ...          0.5
2956    152  ...          NaN

[2 rows x 62 columns]

None


울산을 군구 기상정보까지 불러올 수 있는 게 아니면, 굳이 군구로 나눌 필요는 없지만 데이터 분석의 편의상, ~도로 된 행정지역(시도)의 경우 시군으로 나눠야 할 것으로 사료됨

연별 데이터로는 해당되지 않는 관광지도 연월별 데이터로 확인하면 결측치가 있는 경우가 있음

In [ ]:
clear_Ycol_idx = []
clear_YMcol_idx = []

for all in range(Check_info.shape[1]):
  # 연별 데이터가 있는 경우와 연월별 데이터가 있는 경우가 이어지는 가정이 아니기 때문에, if~elif~가 아닌 if~ if~로 조건반복문 작성
  if all not in del_Ycol_idx:
    clear_Ycol_idx.append(all)

  if all not in del_YMcol_idx:
    clear_YMcol_idx.append(all)

In [ ]:
print(f"3개년 사용가능한 관광지 수 : {len(clear_Ycol_idx)}")
print(f"36개월 사용가능한 관광지 수 : {len(clear_YMcol_idx)}")

3개년 사용가능한 관광지 수 : 30
36개월 사용가능한 관광지 수 : 24


In [ ]:
YR_Ulsan = Check_YEAR.iloc[:, clear_Ycol_idx]
print(YR_Ulsan.shape)
YR_Ulsan

(9, 30)


2      3       4        5   ...     56      57     58     61
2011년  74971  32227     NaN      NaN  ...  46927  102915    NaN  49243
2012년  82505  42306     NaN      NaN  ...  38492   92190  14869  47631
2013년    NaN    NaN     NaN      NaN  ...  51454     NaN  53668    NaN
2014년    NaN    NaN     NaN      NaN  ...  45354   94168  28660  32585
2015년    NaN    NaN     NaN      NaN  ...  49603   80501  14587  19791
2016년    NaN    NaN     NaN      NaN  ...  46049  159065  15598  35875
2017년  45212  46460  183989  1062769  ...  56473   85259  11737  41918
2018년  46184  42413  116856   905378  ...  44329   88751   7062  39982
2019년  50918  48822  153318   826651  ...  49923  100244  14263  31490

[9 rows x 30 columns]

In [ ]:
YM_Ulsan = Check_YM.iloc[:, clear_YMcol_idx]

In [ ]:
YM_Ulsan.shape

(108, 24)

In [ ]:
YM_Ulsan

2     3      4       5      6   ...    47      52    56     57    58
2011년 01월   6107  2727    NaN     NaN    NaN  ...   NaN   50378  2960   6537   NaN
2011년 02월   5920  1696    NaN     NaN    NaN  ...   NaN   24037  3852   6210   NaN
2011년 03월   6617  2924    NaN     NaN    NaN  ...   NaN   21380  3970   7081   NaN
2011년 04월   7115  3339    NaN     NaN    NaN  ...   NaN   45856  4922  11598   NaN
2011년 05월   7173  3472    NaN     NaN    NaN  ...   NaN   29961  6361  11165   NaN
...          ...   ...    ...     ...    ...  ...   ...     ...   ...    ...   ...
2019년 08월   3194  3835  13612  146176   7020  ...  9098  132749  4618  10059  1863
2019년 09월   4208  4133  14490  143156  10431  ...  4286   28088  5674   7005  1241
2019년 10월  11857  7722  13564  126882  12761  ...  8399   35560  4902  10255  1908
2019년 11월   3685  4358  13117   99404   8040  ...  9853   25555  4852   9387  1239
2019년 12월   2392  3429   8637   46747   5928  ...  6891   27303  2360   7178  1339

[108 rows x 24 columns]

### 광주 기상대>광주

## re-ASOS(YMD)

In [ ]:
for area in darea_lst:
  print(f"\n\n{area}_WT_{dtime_lst[2]}\n")
  print(locals()[f"{area}_WT_{dtime_lst[2]}"].shape)
  print(locals()[f"{area}_WT_{dtime_lst[2]}"].columns)
  print(display(locals()[f"{area}_WT_{dtime_lst[2]}"].head(2)))



Ulsan_WT_YMD

(3287, 62)
Index(['지점', '지점명', '일시', '평균기온(°C)', '최저기온(°C)', '최저기온 시각(hhmi)', '최고기온(°C)',
       '최고기온 시각(hhmi)', '강수 계속시간(hr)', '10분 최다 강수량(mm)', '10분 최다강수량 시각(hhmi)',
       '1시간 최다강수량(mm)', '1시간 최다 강수량 시각(hhmi)', '일강수량(mm)', '최대 순간 풍속(m/s)',
       '최대 순간 풍속 풍향(16방위)', '최대 순간풍속 시각(hhmi)', '최대 풍속(m/s)', '최대 풍속 풍향(16방위)',
       '최대 풍속 시각(hhmi)', '평균 풍속(m/s)', '풍정합(100m)', '최다풍향(16방위)',
       '평균 이슬점온도(°C)', '최소 상대습도(%)', '최소 상대습도 시각(hhmi)', '평균 상대습도(%)',
       '평균 증기압(hPa)', '평균 현지기압(hPa)', '최고 해면기압(hPa)', '최고 해면기압 시각(hhmi)',
       '최저 해면기압(hPa)', '최저 해면기압 시각(hhmi)', '평균 해면기압(hPa)', '가조시간(hr)',
       '합계 일조시간(hr)', '1시간 최다일사 시각(hhmi)', '1시간 최다일사량(MJ/m2)', '합계 일사량(MJ/m2)',
       '일 최심신적설(cm)', '일 최심신적설 시각(hhmi)', '일 최심적설(cm)', '일 최심적설 시각(hhmi)',
       '합계 3시간 신적설(cm)', '평균 전운량(1/10)', '평균 중하층운량(1/10)', '평균 지면온도(°C)',
       '최저 초상온도(°C)', '평균 5cm 지중온도(°C)', '평균 10cm 지중온도(°C)',
       '평균 20cm 지중온도(°C)', '평균 30cm 지중온도(°C)', '0.5m 지중온도(°C)',
       '1.0m 지중온도(°C)',

지점 지점명          일시  평균기온(°C)  ...  합계 소형증발량(mm)  9-9강수(mm)   기사  안개 계속시간(hr)
0  152  울산  2011-01-01      -1.4  ...           NaN        NaN  NaN          NaN
1  152  울산  2011-01-02       1.8  ...           NaN        NaN  NaN          NaN

[2 rows x 62 columns]

None


Guangju_WT_YMD

(3287, 62)
Index(['지점', '지점명', '일시', '평균기온(°C)', '최저기온(°C)', '최저기온 시각(hhmi)', '최고기온(°C)',
       '최고기온 시각(hhmi)', '강수 계속시간(hr)', '10분 최다 강수량(mm)', '10분 최다강수량 시각(hhmi)',
       '1시간 최다강수량(mm)', '1시간 최다 강수량 시각(hhmi)', '일강수량(mm)', '최대 순간 풍속(m/s)',
       '최대 순간 풍속 풍향(16방위)', '최대 순간풍속 시각(hhmi)', '최대 풍속(m/s)', '최대 풍속 풍향(16방위)',
       '최대 풍속 시각(hhmi)', '평균 풍속(m/s)', '풍정합(100m)', '최다풍향(16방위)',
       '평균 이슬점온도(°C)', '최소 상대습도(%)', '최소 상대습도 시각(hhmi)', '평균 상대습도(%)',
       '평균 증기압(hPa)', '평균 현지기압(hPa)', '최고 해면기압(hPa)', '최고 해면기압 시각(hhmi)',
       '최저 해면기압(hPa)', '최저 해면기압 시각(hhmi)', '평균 해면기압(hPa)', '가조시간(hr)',
       '합계 일조시간(hr)', '1시간 최다일사 시각(hhmi)', '1시간 최다일사량(MJ/m2)', '합계 일사량(MJ/m2)',
       '일 최심신적설(cm)', '일 최심신적설 시각(hhmi)', '일 최심적설(cm)', '일 최심적설 시각(hhmi)',
       '합계 3시간 신적설(cm)', '평균 전운량(1/10)', '평균 중하층운량(1/10)', '평균 지면온도(°C)',
       '최저 초상온도(°C)', '평균 5cm 지중온도(°C)', '평균 10cm 지중온도(°C)',
       '평균 20cm 지중온도(°C)', '평균 30cm 지중온도(°C)', '0.5m 지중온도(°C)',
       '1.0m 지중온

지점 지점명          일시  ...  9-9강수(mm)                            기사  안개 계속시간(hr)
0  156  광주  2011-01-01  ...        NaN               {소낙눈}0155-0218.          NaN
1  156  광주  2011-01-02  ...        NaN  {박무}0425-{박무}{강도0}0600-0750.          NaN

[2 rows x 62 columns]

None


Gangwon_WT_YMD

(43917, 62)
Index(['지점', '지점명', '일시', '평균기온(°C)', '최저기온(°C)', '최저기온 시각(hhmi)', '최고기온(°C)',
       '최고기온 시각(hhmi)', '강수 계속시간(hr)', '10분 최다 강수량(mm)', '10분 최다강수량 시각(hhmi)',
       '1시간 최다강수량(mm)', '1시간 최다 강수량 시각(hhmi)', '일강수량(mm)', '최대 순간 풍속(m/s)',
       '최대 순간 풍속 풍향(16방위)', '최대 순간풍속 시각(hhmi)', '최대 풍속(m/s)', '최대 풍속 풍향(16방위)',
       '최대 풍속 시각(hhmi)', '평균 풍속(m/s)', '풍정합(100m)', '최다풍향(16방위)',
       '평균 이슬점온도(°C)', '최소 상대습도(%)', '최소 상대습도 시각(hhmi)', '평균 상대습도(%)',
       '평균 증기압(hPa)', '평균 현지기압(hPa)', '최고 해면기압(hPa)', '최고 해면기압 시각(hhmi)',
       '최저 해면기압(hPa)', '최저 해면기압 시각(hhmi)', '평균 해면기압(hPa)', '가조시간(hr)',
       '합계 일조시간(hr)', '1시간 최다일사 시각(hhmi)', '1시간 최다일사량(MJ/m2)', '합계 일사량(MJ/m2)',
       '일 최심신적설(cm)', '일 최심신적설 시각(hhmi)', '일 최심적설(cm)', '일 최심적설 시각(hhmi)',
       '합계 3시간 신적설(cm)', '평균 전운량(1/10)', '평균 중하층운량(1/10)', '평균 지면온도(°C)',
       '최저 초상온도(°C)', '평균 5cm 지중온도(°C)', '평균 10cm 지중온도(°C)',
       '평균 20cm 지중온도(°C)', '평균 30cm 지중온도(°C)', '0.5m 지중온도(°C)',
       '1.0m 지중

지점 지점명  ...                                                 기사  안개 계속시간(hr)
0  90  속초  ...  {눈}0540-{눈}{강도0}0600-{눈}{강도0}0900-{눈}{강도0}1200...          NaN
1  90  속초  ...      {박무}2310-{박무}{강도0}2400- {눈}2330-{눈}{강도0}2400-          NaN

[2 rows x 62 columns]

None


Gyeonggi_WT_YMD

(16435, 62)
Index(['지점', '지점명', '일시', '평균기온(°C)', '최저기온(°C)', '최저기온 시각(hhmi)', '최고기온(°C)',
       '최고기온 시각(hhmi)', '강수 계속시간(hr)', '10분 최다 강수량(mm)', '10분 최다강수량 시각(hhmi)',
       '1시간 최다강수량(mm)', '1시간 최다 강수량 시각(hhmi)', '일강수량(mm)', '최대 순간 풍속(m/s)',
       '최대 순간 풍속 풍향(16방위)', '최대 순간풍속 시각(hhmi)', '최대 풍속(m/s)', '최대 풍속 풍향(16방위)',
       '최대 풍속 시각(hhmi)', '평균 풍속(m/s)', '풍정합(100m)', '최다풍향(16방위)',
       '평균 이슬점온도(°C)', '최소 상대습도(%)', '최소 상대습도 시각(hhmi)', '평균 상대습도(%)',
       '평균 증기압(hPa)', '평균 현지기압(hPa)', '최고 해면기압(hPa)', '최고 해면기압 시각(hhmi)',
       '최저 해면기압(hPa)', '최저 해면기압 시각(hhmi)', '평균 해면기압(hPa)', '가조시간(hr)',
       '합계 일조시간(hr)', '1시간 최다일사 시각(hhmi)', '1시간 최다일사량(MJ/m2)', '합계 일사량(MJ/m2)',
       '일 최심신적설(cm)', '일 최심신적설 시각(hhmi)', '일 최심적설(cm)', '일 최심적설 시각(hhmi)',
       '합계 3시간 신적설(cm)', '평균 전운량(1/10)', '평균 중하층운량(1/10)', '평균 지면온도(°C)',
       '최저 초상온도(°C)', '평균 5cm 지중온도(°C)', '평균 10cm 지중온도(°C)',
       '평균 20cm 지중온도(°C)', '평균 30cm 지중온도(°C)', '0.5m 지중온도(°C)',
       '1.0m 지

지점  지점명  ...                                                 기사  안개 계속시간(hr)
0  98  동두천  ...                            {박무}2220-{박무}{강도0}2400-          NaN
1  98  동두천  ...  -{박무}-{박무}{강도0}0300-{박무}{강도0}0600-{박무}{강도0}090...          NaN

[2 rows x 62 columns]

None


In [ ]:
for area in darea_lst:
  print(f"\n\n{area}_WT_{dtime_lst[2]}\n")
  print(locals()[f"{area}_WT_{dtime_lst[2]}"].info())
  print(locals()[f"{area}_WT_{dtime_lst[2]}"].describe())



Ulsan_WT_YMD

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3287 entries, 0 to 3286
Data columns (total 62 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   지점                   3287 non-null   int64  
 1   지점명                  3287 non-null   object 
 2   일시                   3287 non-null   object 
 3   평균기온(°C)             3287 non-null   float64
 4   최저기온(°C)             3287 non-null   float64
 5   최저기온 시각(hhmi)        3287 non-null   int64  
 6   최고기온(°C)             3287 non-null   float64
 7   최고기온 시각(hhmi)        3287 non-null   int64  
 8   강수 계속시간(hr)          1188 non-null   float64
 9   10분 최다 강수량(mm)       788 non-null    float64
 10  10분 최다강수량 시각(hhmi)   524 non-null    float64
 11  1시간 최다강수량(mm)        788 non-null    float64
 12  1시간 최다 강수량 시각(hhmi)  551 non-null    float64
 13  일강수량(mm)             1188 non-null   float64
 14  최대 순간 풍속(m/s)        3287 non-null   float64
 15  최대 순간 풍속 풍향(16방위)    3

In [ ]:
US_unq = Ulsan_WT_YMD["지점명"].unique()
print("울산", len(US_unq))
GJ_unq = Guangju_WT_YMD["지점명"].unique()
print("광주", len(GJ_unq))

울산 1
광주 1


In [ ]:
GGDo_unq = Gyeonggi_WT_YMD["지점명"].unique()
print("경기도", len(GGDo_unq))
GWDo_unq = Gangwon_WT_YMD["지점명"].unique()
print("강원도", len(GWDo_unq))

경기도 5
강원도 14


In [ ]:
GG_During = YMDduring*5
GW_During = YMDduring*14

In [ ]:
for code, area in enumerate(darea_lst):
  # print(f"\n\n{area}_WT_{dtime_lst[2]}\n")
  if code<=1:
    if locals()[f"{area}_WT_{dtime_lst[2]}"].shape[0]==YMDduring:
      print(f"\n{area}_WT_{dtime_lst[2]}에는 기상 결측일이 없습니다.")
    else:
      print(f"\n{area}_WT_{dtime_lst[2]}에는 기상 결측일이 존재합니다.")
  elif code==2:
    if locals()[f"{area}_WT_{dtime_lst[2]}"].shape[0]==GW_During:
      print(f"\n{area}_WT_{dtime_lst[2]}에는 기상 결측일이 없습니다.")
    else:
      print(f"\n{area}_WT_{dtime_lst[2]}에는 기상 결측일이 존재합니다.")
  else:
    if locals()[f"{area}_WT_{dtime_lst[2]}"].shape[0]==GG_During:
      print(f"\n{area}_WT_{dtime_lst[2]}에는 기상 결측일이 없습니다.")
    else:
      print(f"\n{area}_WT_{dtime_lst[2]}에는 기상 결측일이 존재합니다.")      


Ulsan_WT_YMD에는 기상 결측일이 존재합니다.

Guangju_WT_YMD에는 기상 결측일이 존재합니다.

Gangwon_WT_YMD에는 기상 결측일이 존재합니다.

Gyeonggi_WT_YMD에는 기상 결측일이 존재합니다.


<big>기상청에서 CSV로 다운받은 데이터도 <b>결측일</b>이 존재함</big>

## re-ASOS(YM)

In [ ]:
for area in darea_lst:
  print(f"\n\n{area}_WT_{dtime_lst[1]}\n")
  print(locals()[f"{area}_WT_{dtime_lst[1]}"].shape)
  print(locals()[f"{area}_WT_{dtime_lst[1]}"].columns)
  print(display(locals()[f"{area}_WT_{dtime_lst[1]}"].head(2)))



Ulsan_WT_YM

(108, 69)
Index(['지점', '지점명', '일시', '평균기온(°C)', '평균최고기온(°C)', '평균최저기온(°C)', '최고기온(°C)',
       '최저기온(°C)', '최고기온 나타난날(yyyymmdd)', '최저기온 나타난날(yyyymmdd)', '평균현지기압(hPa)',
       '평균해면기압(hPa)', '최고해면기압(hPa)', '최저해면기압(hPa)', '최고해면기압 나타난날(yyyymmdd)',
       '최저해면기압 나타난날(yyyymmdd)', '평균수증기압(hPa)', '최고수증기압(hPa)', '최저수증기압(hPa)',
       '최고수증기압 나타난날(yyyymmdd)', '최저수증기압 나타난날(yyyymmdd)', '평균이슬점온도(°C)',
       '평균상대습도(%)', '최소상대습도(%)', '최소상대습도 나타난날(yyyymmdd)', '월합강수량(00~24h만)(mm)',
       '일최다강수량(mm)', '1시간최다강수량(mm)', '10분최다강수량(mm)', '일최다강수량 나타난날(yyyymmdd)',
       '1시간최다강수량 나타난날(yyyymmdd)', '10분최다강수량 나타난날(yyyymmdd)', '소형총증발량(mm)',
       '소형일최대증발량(mm)', '대형총증발량(mm)', '대형일최대증발량(mm)', '소형일최대증발량 나타난날(yyyymmdd)',
       '대형일최대증발량 나타난날(yyyymmdd)', '평균풍속(m/s)', '최대풍속(m/s)', '최대순간풍속(m/s)',
       '최대풍속 풍향(16방위)', '최대순간풍속 풍향(16방위)', '최대풍속 나타난날(yyyymmdd)',
       '최대순간풍속 나타난날(yyyymmdd)', '최다풍향(16방위)', '평균운량(1/10)', '평균중하층운량(1/10)',
       '합계 일조시간(hr)', '일조율(%)', '합계 일사량(MJ/m2)', '최심적설(cm)',

지점 지점명       일시  ...  1.5m평균지중온도(°C)  3.0m평균지중온도(°C)  5.0m평균지중온도(°C)
0  152  울산  2011-01  ...             NaN             NaN             NaN
1  152  울산  2011-02  ...             NaN             NaN             NaN

[2 rows x 69 columns]

None


Guangju_WT_YM

(108, 69)
Index(['지점', '지점명', '일시', '평균기온(°C)', '평균최고기온(°C)', '평균최저기온(°C)', '최고기온(°C)',
       '최저기온(°C)', '최고기온 나타난날(yyyymmdd)', '최저기온 나타난날(yyyymmdd)', '평균현지기압(hPa)',
       '평균해면기압(hPa)', '최고해면기압(hPa)', '최저해면기압(hPa)', '최고해면기압 나타난날(yyyymmdd)',
       '최저해면기압 나타난날(yyyymmdd)', '평균수증기압(hPa)', '최고수증기압(hPa)', '최저수증기압(hPa)',
       '최고수증기압 나타난날(yyyymmdd)', '최저수증기압 나타난날(yyyymmdd)', '평균이슬점온도(°C)',
       '평균상대습도(%)', '최소상대습도(%)', '최소상대습도 나타난날(yyyymmdd)', '월합강수량(00~24h만)(mm)',
       '일최다강수량(mm)', '1시간최다강수량(mm)', '10분최다강수량(mm)', '일최다강수량 나타난날(yyyymmdd)',
       '1시간최다강수량 나타난날(yyyymmdd)', '10분최다강수량 나타난날(yyyymmdd)', '소형총증발량(mm)',
       '소형일최대증발량(mm)', '대형총증발량(mm)', '대형일최대증발량(mm)', '소형일최대증발량 나타난날(yyyymmdd)',
       '대형일최대증발량 나타난날(yyyymmdd)', '평균풍속(m/s)', '최대풍속(m/s)', '최대순간풍속(m/s)',
       '최대풍속 풍향(16방위)', '최대순간풍속 풍향(16방위)', '최대풍속 나타난날(yyyymmdd)',
       '최대순간풍속 나타난날(yyyymmdd)', '최다풍향(16방위)', '평균운량(1/10)', '평균중하층운량(1/10)',
       '합계 일조시간(hr)', '일조율(%)', '합계 일사량(MJ/m2)', '최심적

지점 지점명       일시  ...  1.5m평균지중온도(°C)  3.0m평균지중온도(°C)  5.0m평균지중온도(°C)
0  156  광주  2011-01  ...             NaN             NaN             NaN
1  156  광주  2011-02  ...             NaN             NaN             NaN

[2 rows x 69 columns]

None


Gangwon_WT_YM

(1443, 69)
Index(['지점', '지점명', '일시', '평균기온(°C)', '평균최고기온(°C)', '평균최저기온(°C)', '최고기온(°C)',
       '최저기온(°C)', '최고기온 나타난날(yyyymmdd)', '최저기온 나타난날(yyyymmdd)', '평균현지기압(hPa)',
       '평균해면기압(hPa)', '최고해면기압(hPa)', '최저해면기압(hPa)', '최고해면기압 나타난날(yyyymmdd)',
       '최저해면기압 나타난날(yyyymmdd)', '평균수증기압(hPa)', '최고수증기압(hPa)', '최저수증기압(hPa)',
       '최고수증기압 나타난날(yyyymmdd)', '최저수증기압 나타난날(yyyymmdd)', '평균이슬점온도(°C)',
       '평균상대습도(%)', '최소상대습도(%)', '최소상대습도 나타난날(yyyymmdd)', '월합강수량(00~24h만)(mm)',
       '일최다강수량(mm)', '1시간최다강수량(mm)', '10분최다강수량(mm)', '일최다강수량 나타난날(yyyymmdd)',
       '1시간최다강수량 나타난날(yyyymmdd)', '10분최다강수량 나타난날(yyyymmdd)', '소형총증발량(mm)',
       '소형일최대증발량(mm)', '대형총증발량(mm)', '대형일최대증발량(mm)', '소형일최대증발량 나타난날(yyyymmdd)',
       '대형일최대증발량 나타난날(yyyymmdd)', '평균풍속(m/s)', '최대풍속(m/s)', '최대순간풍속(m/s)',
       '최대풍속 풍향(16방위)', '최대순간풍속 풍향(16방위)', '최대풍속 나타난날(yyyymmdd)',
       '최대순간풍속 나타난날(yyyymmdd)', '최다풍향(16방위)', '평균운량(1/10)', '평균중하층운량(1/10)',
       '합계 일조시간(hr)', '일조율(%)', '합계 일사량(MJ/m2)', '최심

지점 지점명       일시  ...  1.5m평균지중온도(°C)  3.0m평균지중온도(°C)  5.0m평균지중온도(°C)
0  90  속초  2011-01  ...             NaN             NaN             NaN
1  90  속초  2011-02  ...             NaN             NaN             NaN

[2 rows x 69 columns]

None


Gyeonggi_WT_YM

(540, 69)
Index(['지점', '지점명', '일시', '평균기온(°C)', '평균최고기온(°C)', '평균최저기온(°C)', '최고기온(°C)',
       '최저기온(°C)', '최고기온 나타난날(yyyymmdd)', '최저기온 나타난날(yyyymmdd)', '평균현지기압(hPa)',
       '평균해면기압(hPa)', '최고해면기압(hPa)', '최저해면기압(hPa)', '최고해면기압 나타난날(yyyymmdd)',
       '최저해면기압 나타난날(yyyymmdd)', '평균수증기압(hPa)', '최고수증기압(hPa)', '최저수증기압(hPa)',
       '최고수증기압 나타난날(yyyymmdd)', '최저수증기압 나타난날(yyyymmdd)', '평균이슬점온도(°C)',
       '평균상대습도(%)', '최소상대습도(%)', '최소상대습도 나타난날(yyyymmdd)', '월합강수량(00~24h만)(mm)',
       '일최다강수량(mm)', '1시간최다강수량(mm)', '10분최다강수량(mm)', '일최다강수량 나타난날(yyyymmdd)',
       '1시간최다강수량 나타난날(yyyymmdd)', '10분최다강수량 나타난날(yyyymmdd)', '소형총증발량(mm)',
       '소형일최대증발량(mm)', '대형총증발량(mm)', '대형일최대증발량(mm)', '소형일최대증발량 나타난날(yyyymmdd)',
       '대형일최대증발량 나타난날(yyyymmdd)', '평균풍속(m/s)', '최대풍속(m/s)', '최대순간풍속(m/s)',
       '최대풍속 풍향(16방위)', '최대순간풍속 풍향(16방위)', '최대풍속 나타난날(yyyymmdd)',
       '최대순간풍속 나타난날(yyyymmdd)', '최다풍향(16방위)', '평균운량(1/10)', '평균중하층운량(1/10)',
       '합계 일조시간(hr)', '일조율(%)', '합계 일사량(MJ/m2)', '최심

지점  지점명       일시  ...  1.5m평균지중온도(°C)  3.0m평균지중온도(°C)  5.0m평균지중온도(°C)
0  98  동두천  2011-01  ...             NaN             NaN             NaN
1  98  동두천  2011-02  ...             NaN             NaN             NaN

[2 rows x 69 columns]

None


In [ ]:
for area in darea_lst:
  print(f"\n\n{area}_WT_{dtime_lst[1]}\n")
  print(locals()[f"{area}_WT_{dtime_lst[1]}"].info())
  print(locals()[f"{area}_WT_{dtime_lst[1]}"].describe())



Ulsan_WT_YM

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 108 entries, 0 to 107
Data columns (total 69 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   지점                       108 non-null    int64  
 1   지점명                      108 non-null    object 
 2   일시                       108 non-null    object 
 3   평균기온(°C)                 108 non-null    float64
 4   평균최고기온(°C)               108 non-null    float64
 5   평균최저기온(°C)               108 non-null    float64
 6   최고기온(°C)                 108 non-null    float64
 7   최저기온(°C)                 108 non-null    float64
 8   최고기온 나타난날(yyyymmdd)      108 non-null    int64  
 9   최저기온 나타난날(yyyymmdd)      108 non-null    int64  
 10  평균현지기압(hPa)              108 non-null    float64
 11  평균해면기압(hPa)              108 non-null    float64
 12  최고해면기압(hPa)              108 non-null    float64
 13  최저해면기압(hPa)              108 non-null    float64
 14  최고해면기압 나타난날

## re-ASOS(Y)

In [ ]:
for area in darea_lst:
  print(f"\n\n{area}_WT_{dtime_lst[0]}\n")
  print(locals()[f"{area}_WT_{dtime_lst[0]}"].shape)
  print(locals()[f"{area}_WT_{dtime_lst[0]}"].columns)
  print(display(locals()[f"{area}_WT_{dtime_lst[0]}"].head(2)))



Ulsan_WT_Year

(9, 22)
Index(['지점', '지점명', '일시', '평균기온(°C)', '최저기온(°C)', '최저기온 나타난날(yyyymmdd)',
       '최고기온(°C)', '최고기온 나타난날(yyyymmdd)', '평균최저기온(°C)', '평균최고기온(°C)',
       '평균 해면기압(hPa)', '합계 강수량(mm)', '일 최다 강수량(mm)', '일 최다 강수량 나타난날(yyyymmdd)',
       '평균 상대습도(%)', '합계 일조시간(hr)', '일조율(%)', '합계 일사량(MJ/m2)', '평균 풍속(m/s)',
       '최다 풍향(16방위)', '합계 소형 증발량(mm)', '합계 대형 증발량(mm)'],
      dtype='object')


지점 지점명    일시  ...  최다 풍향(16방위)  합계 소형 증발량(mm)  합계 대형 증발량(mm)
0  152  울산  2011  ...          320            NaN            NaN
1  152  울산  2012  ...           20            NaN            NaN

[2 rows x 22 columns]

None


Guangju_WT_Year

(9, 22)
Index(['지점', '지점명', '일시', '평균기온(°C)', '최저기온(°C)', '최저기온 나타난날(yyyymmdd)',
       '최고기온(°C)', '최고기온 나타난날(yyyymmdd)', '평균최저기온(°C)', '평균최고기온(°C)',
       '평균 해면기압(hPa)', '합계 강수량(mm)', '일 최다 강수량(mm)', '일 최다 강수량 나타난날(yyyymmdd)',
       '평균 상대습도(%)', '합계 일조시간(hr)', '일조율(%)', '합계 일사량(MJ/m2)', '평균 풍속(m/s)',
       '최다 풍향(16방위)', '합계 소형 증발량(mm)', '합계 대형 증발량(mm)'],
      dtype='object')


지점 지점명    일시  ...  최다 풍향(16방위)  합계 소형 증발량(mm)  합계 대형 증발량(mm)
0  156  광주  2011  ...           20            NaN            NaN
1  156  광주  2012  ...           50            NaN            NaN

[2 rows x 22 columns]

None


Gangwon_WT_Year

(121, 22)
Index(['지점', '지점명', '일시', '평균기온(°C)', '최저기온(°C)', '최저기온 나타난날(yyyymmdd)',
       '최고기온(°C)', '최고기온 나타난날(yyyymmdd)', '평균최저기온(°C)', '평균최고기온(°C)',
       '평균 해면기압(hPa)', '합계 강수량(mm)', '일 최다 강수량(mm)', '일 최다 강수량 나타난날(yyyymmdd)',
       '평균 상대습도(%)', '합계 일조시간(hr)', '일조율(%)', '합계 일사량(MJ/m2)', '평균 풍속(m/s)',
       '최다 풍향(16방위)', '합계 소형 증발량(mm)', '합계 대형 증발량(mm)'],
      dtype='object')


지점 지점명    일시  ...  최다 풍향(16방위)  합계 소형 증발량(mm)  합계 대형 증발량(mm)
0  90  속초  2011  ...          290            NaN            NaN
1  90  속초  2012  ...          250            NaN            NaN

[2 rows x 22 columns]

None


Gyeonggi_WT_Year

(45, 22)
Index(['지점', '지점명', '일시', '평균기온(°C)', '최저기온(°C)', '최저기온 나타난날(yyyymmdd)',
       '최고기온(°C)', '최고기온 나타난날(yyyymmdd)', '평균최저기온(°C)', '평균최고기온(°C)',
       '평균 해면기압(hPa)', '합계 강수량(mm)', '일 최다 강수량(mm)', '일 최다 강수량 나타난날(yyyymmdd)',
       '평균 상대습도(%)', '합계 일조시간(hr)', '일조율(%)', '합계 일사량(MJ/m2)', '평균 풍속(m/s)',
       '최다 풍향(16방위)', '합계 소형 증발량(mm)', '합계 대형 증발량(mm)'],
      dtype='object')


지점  지점명    일시  ...  최다 풍향(16방위)  합계 소형 증발량(mm)  합계 대형 증발량(mm)
0  98  동두천  2011  ...          360            NaN            NaN
1  98  동두천  2012  ...          360            NaN            NaN

[2 rows x 22 columns]

None


In [ ]:
for area in darea_lst:
  print(f"\n\n{area}_WT_{dtime_lst[0]}\n")
  print(locals()[f"{area}_WT_{dtime_lst[0]}"].info())
  print(locals()[f"{area}_WT_{dtime_lst[0]}"].describe())



Ulsan_WT_Year

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9 entries, 0 to 8
Data columns (total 22 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   지점                       9 non-null      int64  
 1   지점명                      9 non-null      object 
 2   일시                       9 non-null      int64  
 3   평균기온(°C)                 9 non-null      float64
 4   최저기온(°C)                 9 non-null      float64
 5   최저기온 나타난날(yyyymmdd)      9 non-null      int64  
 6   최고기온(°C)                 9 non-null      float64
 7   최고기온 나타난날(yyyymmdd)      9 non-null      int64  
 8   평균최저기온(°C)               9 non-null      float64
 9   평균최고기온(°C)               9 non-null      float64
 10  평균 해면기압(hPa)             9 non-null      float64
 11  합계 강수량(mm)               9 non-null      float64
 12  일 최다 강수량(mm)             9 non-null      float64
 13  일 최다 강수량 나타난날(yyyymmdd)  9 non-null      int64  
 14  평균 상대습도(%)   

## 기상 데이터 결측일 원인

* 종관기상관측 장비로 관측한 일 기상자료를 조회하는 서비스
* 기상 환경 및 신호 상태에 따라 관측이 잘 이뤄질 때도 아닐 때도 있음
* 추가 관측지 생성 여부에 따라, 그 이전의 결측치를 구할 방법이 없음
* 행정 구역이 변한 경우, 결측 데이터 많을 수 밖에 없음

* 2년씩 불러올 때 에러 발견_에러발생시기 및 지점(관측 시기 조회 필요)
  * 2011년~2014년도 북춘천지점
  * 2017년~2018년도 대구지점
  * 2011년~2014년도 홍성지점
  * 2011년~2018년도 세종지점
  * 2017년~2018년도 남해지점
  * 2011년~2012년도 진도군지점


## Visitors

In [ ]:
def data_load_sep_local(path, name):
  #데이터 불러오고 내국인만 추출한 뒤, 컬럼 바꿔주기
  #path엔 "주요관광지점입장객데이터 경로 및 파일명, 확장자명" 선언
  #name : 해당 데이터를 저장할 변수명(예. "GangWon")

  #데이터 불러오기  
  locals()[f"{name}"] = pd.read_excel(path)
  #내국인만 추출한 뒤 저장
  locals()[f"{name}_local"] = locals()[f"{name}"][locals()[f"{name}"].iloc[:,3]=="내국인"]
  
  #년월이 "Unnamed:n" 으로 저장되어서 그냥 한번에 바꿈
  #저장될 필요가 없는 변수는 locals로 불러오지 않음
  dataCol = [col for col in locals()[f"{name}_local"].columns]
  dataNCol = [ncol for ncol in locals()[f"{name}"].iloc[0,:]]
  dataFCol = {}
  for og, nw in zip(dataCol, dataNCol):
    if ":" in og:
      dataFCol[og]=nw
    else:
      dataFCol[og]=og
  locals()[f"{name}_local"].rename(columns=dataFCol, inplace=True)
  
  #내국인 추출로 난장난 index 초기화
  locals()[f"{name}_local"].reset_index(drop=True, inplace=True)
  
  return locals()[f"{name}"], locals()[f"{name}_local"]

def sep_sgg(local_df, initial):
  #시군구로 지역범위가 넓은 곳의 데이터 분류
  #local_df : 위의 함수로 불러온 내국인만 있는 데이터
  #initial : 시군구별 분류한 데이터를 저장할 이니셜(예. "GW_춘천시" 중 "GW")
  
  globals()[f"{initial}_SGG"] = [sgg for sgg in local_df.iloc[:,1].unique()]

  for sgg in globals()[f"{initial}_SGG"]:
    globals()[f"{initial}_{sgg}"] = local_df[local_df.iloc[:,1]==sgg]
    globals()[f"{initial}_{sgg}"].reset_index(drop=True, inplace=True)
    print(f"'{initial}_{sgg}'의 shape은 ", globals()[f"{initial}_{sgg}"].shape, "입니다.")
  #globals를 활용해 앞에 변수 지정 없이도 불러올 수 있게 설정함

### 울산 관광지 입장객

In [ ]:
path = f"{CHK_PATH}/울산_주요관광지점_입장객_2011_2019.xls"
name = darea_lst[0]

UlSan_df, UlSan_local = data_load_sep_local(path, name)

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4308: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [ ]:
del_Ycol_idx = []   #연별 데이터 중 결측치가 존재하는 관광지 입장객
del_YMcol_idx = []  #연월별 데이터 중 결측치가 존재하는 관광지 입장객

for check in range(Check_info.shape[1]):
  for TorF, TorF2 in zip(Check_YEAR.iloc[-3:, check].isna(), Check_YEAR.iloc[-3:, check].isin([0])):
    if TorF==True:
      del_Ycol_idx.append(check)
    if TorF2==True: #NaN이 아닌 int(0)으로 입력된 결측치도 있었음
      del_Ycol_idx.append(check)
    
  for TorF, TorF2 in zip(Check_YM.iloc[-36:, check].isna(), Check_YM.iloc[-36:, check].isin([0])):
    if TorF==True:
      del_YMcol_idx.append(check)
    if TorF2==True: #NaN이 아닌 int(0)으로 입력된 결측치도 있었음
      del_YMcol_idx.append(check)

del_Ycol_idx = sorted(list(set(del_Ycol_idx)))    #3개년 true값 중복으로 들어갔을 check 넘버 제거
del_YMcol_idx = sorted(list(set(del_YMcol_idx)))  #36개월 true값 중복으로 들어갔을 check 넘버 제거

NameError: ignored

### 광주 관광지 입장객

### 강원도 관광지 입장객

## 기상청_지상(종관, ASOS) 일자료 조회서비스

### 경기도 관광지 입장객

In [ ]:
Check_YM.rename(columns=ncol)
Check_YM

0    1      2     3      4   ...     57    58     59     60    61
2011년 01월  1178  NaN   6107  2727    NaN  ...   6537   NaN   9910  45850  1089
2011년 02월   905  NaN   5920  1696    NaN  ...   6210   NaN   8625  30605  1030
2011년 03월  1840  NaN   6617  2924    NaN  ...   7081   NaN  42550  28210  1435
2011년 04월  1955  NaN   7115  3339    NaN  ...  11598   NaN  77400  28875  5906
2011년 05월  2020  NaN   7173  3472    NaN  ...  11165   NaN  52850  47600  3950
...         ...  ...    ...   ...    ...  ...    ...   ...    ...    ...   ...
2019년 08월   NaN  626   3194  3835  13612  ...  10059  1863    NaN    NaN  5240
2019년 09월   NaN  485   4208  4133  14490  ...   7005  1241    NaN    NaN  4740
2019년 10월   NaN  817  11857  7722  13564  ...  10255  1908    NaN    NaN  4720
2019년 11월   NaN  730   3685  4358  13117  ...   9387  1239    NaN    NaN  2430
2019년 12월   NaN  749   2392  3429   8637  ...   7178  1339    NaN    NaN  1260

[108 rows x 62 columns]

#### 3개년, 36개월 결측치 없는 데이터(YR_Ulsan, YM_Ulsan)

2019년을 기준으로 3개년의 데이터가 존재하는 데이터만 사용하기

In [ ]:
for check in range(Check_info.shape[1]):
  for TorF in Check_YEAR.iloc[-3:, check].isin([0]):
    if TorF==True:
      print(check)
      print(TorF)

55
True


In [ ]:
del_Ycol_idx = []   #연별 데이터 중 결측치가 존재하는 관광지 입장객
del_YMcol_idx = []  #연월별 데이터 중 결측치가 존재하는 관광지 입장객

for check in range(Check_info.shape[1]):
  for TorF, TorF2 in zip(Check_YEAR.iloc[-3:, check].isna(), Check_YEAR.iloc[-3:, check].isin([0])):
    if TorF==True:
      del_Ycol_idx.append(check)
    if TorF2==True: #NaN이 아닌 int(0)으로 입력된 결측치도 있었음
      del_Ycol_idx.append(check)
    
  for TorF, TorF2 in zip(Check_YM.iloc[-36:, check].isna(), Check_YM.iloc[-36:, check].isin([0])):
    if TorF==True:
      del_YMcol_idx.append(check)
    if TorF2==True: #NaN이 아닌 int(0)으로 입력된 결측치도 있었음
      del_YMcol_idx.append(check)

del_Ycol_idx = sorted(list(set(del_Ycol_idx)))    #3개년 true값 중복으로 들어갔을 check 넘버 제거
del_YMcol_idx = sorted(list(set(del_YMcol_idx)))  #36개월 true값 중복으로 들어갔을 check 넘버 제거

In [ ]:
print(f"3개년 결측치 존재하는 관광지 수 : {len(del_Ycol_idx)}")
print(f"36개월 결측치 존재하는 관광지 수 : {len(del_YMcol_idx)}")

3개년 결측치 존재하는 관광지 수 : 32
36개월 결측치 존재하는 관광지 수 : 38


In [ ]:
for area in darea_lst:
  print(f"\n\n{area}_WT_{dtime_lst[1]}\n")
  print(locals()[f"{area}_WT_{dtime_lst[1]}"].shape)
  print(locals()[f"{area}_WT_{dtime_lst[1]}"].columns)
  print(display(locals()[f"{area}_WT_{dtime_lst[1]}"].head(2)))

In [ ]:
for area in darea_lst:
  print(f"\n\n{area}_WT_{dtime_lst[0]}\n")
  print(locals()[f"{area}_WT_{dtime_lst[0]}"].info())
  print(locals()[f"{area}_WT_{dtime_lst[0]}"].describe())

In [ ]:
stD, edD, during = Date_Duration()

# 데이터 분석

## EDA & VSL

## Model

# 데이터 표현

# 개인폴더 내 데이터 처리

## 주요관광지점입장객 데이터에서 사용가능한 함수

In [ ]:
%cd "{CHK_PATH}"
!ls

/content/drive/.shortcut-targets-by-id/1srYR8TqwFsUSR4BYzQtIBmWUKrEBjd44/Task/Proj_WeatherGo/MINE
경기_주요관광지점_입장객_2011_2019.xls	울산_주요관광지점_입장객_2011_2019.xls
광주_주요관광지점_입장객_2011_2019.xls	강원_주요관광지점_입장객_2011_2019.xls


In [ ]:
def data_load_sep_local(path, name):
  #데이터 불러오고 내국인만 추출한 뒤, 컬럼 바꿔주기
  #path엔 "주요관광지점입장객데이터 경로 및 파일명, 확장자명" 선언
  #name : 해당 데이터를 저장할 변수명(예. "GangWon")

  #데이터 불러오기  
  locals()[f"{name}"] = pd.read_excel(path)
  #내국인만 추출한 뒤 저장
  locals()[f"{name}_local"] = locals()[f"{name}"][locals()[f"{name}"].iloc[:,3]=="내국인"]
  
  #년월이 "Unnamed:n" 으로 저장되어서 그냥 한번에 바꿈
  #저장될 필요가 없는 변수는 locals로 불러오지 않음
  dataCol = [col for col in locals()[f"{name}_local"].columns]
  dataNCol = [ncol for ncol in locals()[f"{name}"].iloc[0,:]]
  dataFCol = {}
  for og, nw in zip(dataCol, dataNCol):
    if ":" in og:
      dataFCol[og]=nw
    else:
      dataFCol[og]=og
  locals()[f"{name}_local"].rename(columns=dataFCol, inplace=True)
  
  #내국인 추출로 난장난 index 초기화
  locals()[f"{name}_local"].reset_index(drop=True, inplace=True)
  
  return locals()[f"{name}"], locals()[f"{name}_local"]

def sep_sgg(local_df, initial):
  #시군구로 지역범위가 넓은 곳의 데이터 분류
  #local_df : 위의 함수로 불러온 내국인만 있는 데이터
  #initial : 시군구별 분류한 데이터를 저장할 이니셜(예. "GW_춘천시" 중 "GW")
  
  globals()[f"{initial}_SGG"] = [sgg for sgg in local_df.iloc[:,1].unique()]

  for sgg in globals()[f"{initial}_SGG"]:
    globals()[f"{initial}_{sgg}"] = local_df[local_df.iloc[:,1]==sgg]
    globals()[f"{initial}_{sgg}"].reset_index(drop=True, inplace=True)
    print(f"'{initial}_{sgg}'의 shape은 ", globals()[f"{initial}_{sgg}"].shape, "입니다.")
  #globals를 활용해 앞에 변수 지정 없이도 불러올 수 있게 설정함

### 울산 관광지 입장객

In [ ]:
path = f"{CHK_PATH}/울산_주요관광지점_입장객_2011_2019.xls"
name = "UlSan"

UlSan_df, UlSan_local = data_load_sep_local(path, name)

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4308: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


#### 울산 내국인만 확인

In [ ]:
Check = UlSan_local.T
Check.head(20)

0            1   ...           60         61
시도                          울산광역시        울산광역시  ...        울산광역시      울산광역시
군구                             중구           중구  ...          울주군        울주군
관광지                    수운 최제우 유허지  약사동제방유적전시관   ...      진하 해수욕장  충렬공박제상기념관
내/외국인                         내국인          내국인  ...          내국인        내국인
총계(2011.01 ~ 2019.12)       32701        15424  ...  1.99138e+06     298515
2011년                       16503          NaN  ...      1001997      49243
2011년 01월                    1178          NaN  ...        45850       1089
2011년 02월                     905          NaN  ...        30605       1030
2011년 03월                    1840          NaN  ...        28210       1435
2011년 04월                    1955          NaN  ...        28875       5906
2011년 05월                    2020          NaN  ...        47600       3950
2011년 06월                    1995          NaN  ...        52650       2945
2011년 07월                    1275          NaN  ...       137884       6890
2011년 08월                    1040          NaN  ...       388248       9930
2011년 09월                    1425          NaN  ...        63500       6500
2011년 10월                    1100          NaN  ...        58675       4958
2011년 11월                     912          NaN  ...        60800       1937
2011년 12월                     858          NaN  ...        59100       2673
2012년                       16198          NaN  ...       989378      47631
2012년 01월                    1031          NaN  ...        54132       1113

[20 rows x 62 columns]

In [ ]:
Check.index

Index(['시도', '군구', '관광지', '내/외국인', '총계(2011.01 ~ 2019.12)', '2011년',
       '2011년 01월', '2011년 02월', '2011년 03월', '2011년 04월',
       ...
       '2019년 03월', '2019년 04월', '2019년 05월', '2019년 06월', '2019년 07월',
       '2019년 08월', '2019년 09월', '2019년 10월', '2019년 11월', '2019년 12월'],
      dtype='object', length=122)

In [ ]:
Check.shape

(122, 62)

#### 울산 info(변수명 : Check_info)

index[:5]까지는 "시도", "군구", "관광지", "내/외국인", "총계(2011.01~2019.12)"를 담은 데이터임

index[:3]을 제외하고 필요한지가 의문이지만, 우선 계속 확인해 볼 필요가 있음

In [ ]:
Check_info = Check.iloc[:5,:]

In [ ]:
Check_all = Check.iloc[5:,:]

#### 9개년 중 선별(연별 인원수 : Check_YEAR, 연월별 인원수 데이터:Check_YM)


* 목적1 : 2019년 기준 3개년의 데이터 유무를 바로 확인하기 위함
  * 先 : 연별 인원수 3개년 존재 확인
  * 後 : 연월별 인원수 36개월 존재 확인(기준 : 2017.01~2019.12)
  
* 목적2 : 데이터가 더럽게 들어와서 어차피 분류해야 함

In [ ]:
Check_all.shape

(117, 62)

In [ ]:
13*9

117

In [ ]:
Check_all

0    1      2      3   ...    58      59       60     61
2011년      16503  NaN  74971  32227  ...   NaN  888986  1001997  49243
2011년 01월   1178  NaN   6107   2727  ...   NaN    9910    45850   1089
2011년 02월    905  NaN   5920   1696  ...   NaN    8625    30605   1030
2011년 03월   1840  NaN   6617   2924  ...   NaN   42550    28210   1435
2011년 04월   1955  NaN   7115   3339  ...   NaN   77400    28875   5906
...          ...  ...    ...    ...  ...   ...     ...      ...    ...
2019년 08월    NaN  626   3194   3835  ...  1863     NaN      NaN   5240
2019년 09월    NaN  485   4208   4133  ...  1241     NaN      NaN   4740
2019년 10월    NaN  817  11857   7722  ...  1908     NaN      NaN   4720
2019년 11월    NaN  730   3685   4358  ...  1239     NaN      NaN   2430
2019년 12월    NaN  749   2392   3429  ...  1339     NaN      NaN   1260

[117 rows x 62 columns]

* 관점 포인트 : NaN값이 어떻게 존재하는가??

In [ ]:
Check_all.iloc[[0,13,26],:]

0    1      2      3    4   ...      57     58      59       60     61
2011년  16503  NaN  74971  32227  NaN  ...  102915    NaN  888986  1001997  49243
2012년  16198  NaN  82505  42306  NaN  ...   92190  14869  932252   989378  47631
2013년    NaN  NaN    NaN    NaN  NaN  ...     NaN  53668     NaN      NaN    NaN

[3 rows x 62 columns]

연별과 연월별로 분류해줘야 하는 이유 중 하나

In [ ]:
YEAR = []
YM = []

for idx, cont in zip(Check_all.index, range(Check_all.shape[0])):
  if ("년" in idx) and ("월" in idx):
    YM.append(cont)
  else :
    YEAR.append(cont)

In [ ]:
Check_YEAR = Check_all.iloc[YEAR, :]
Check_YM = Check_all.iloc[YM, :]

In [ ]:
#관광지명으로 칼럼명 만들기
ncol = {}

for tst in Check_info:
  ncol[tst] = str(Check_info.iloc[0, tst])[:2]+"_"+str(Check_info.iloc[1, tst])+"_"+str(Check_info.iloc[2, tst])

In [ ]:
Check_YEAR.rename(columns=ncol)
Check_YEAR

0     1      2      3       4   ...      57     58      59       60     61
2011년  16503   NaN  74971  32227     NaN  ...  102915    NaN  888986  1001997  49243
2012년  16198   NaN  82505  42306     NaN  ...   92190  14869  932252   989378  47631
2013년    NaN   NaN    NaN    NaN     NaN  ...     NaN  53668     NaN      NaN    NaN
2014년    NaN   NaN    NaN    NaN     NaN  ...   94168  28660     NaN      NaN  32585
2015년    NaN   NaN    NaN    NaN     NaN  ...   80501  14587     NaN      NaN  19791
2016년    NaN   NaN    NaN    NaN     NaN  ...  159065  15598     NaN      NaN  35875
2017년    NaN   NaN  45212  46460  183989  ...   85259  11737     NaN      NaN  41918
2018년    NaN  6717  46184  42413  116856  ...   88751   7062     NaN      NaN  39982
2019년    NaN  8707  50918  48822  153318  ...  100244  14263     NaN      NaN  31490

[9 rows x 62 columns]

# 공유폴더 내 데이터 처리

# 공유폴더(주석처리)

## ~~기상청_관광코스별 관광지 상세날씨(우선 주석처리<-이유 : 2011년부터의 데이터 없음)~~

In [ ]:
# SV1 = join(DATA_PATH, "기상청_관광코스별_관광지_상세날씨")
# %cd "{SV1}"
# !ls

/content/drive/.shortcut-targets-by-id/1TVtWvZ4DNn7V20Aj6JzSd5_iRdbjEXbG/Proj_WT/DataSets/기상청_관광코스별_관광지_상세날씨
시군구별관광기후지수_2017.csv  시군구별관광기후지수_2020.csv
시군구별관광기후지수_2018.csv  시군구별관광기후지수_2021.csv
시군구별관광기후지수_2019.csv  시군구별관광기후지수.csv


In [ ]:
# #여러 서비스 중 우선 "시군구별관광기후지수"
# SGG_S_W = "시군구별관광기후지수"

In [ ]:
# for yr in range(2017, 2022):
#   locals()[f'sgg{str(yr)[-2:]}'] = pd.read_csv(f'{SV1}/{SGG_S_W}_{str(yr)}.csv')

In [ ]:
# sgg17.head()

예보시각 전체 도시 이름  시도  시군구       시군구코드  관광기후지수 관광기후지수등급
0  2017-12-31 00:00   서울 강남구  서울  강남구  1168000000   38.15       나쁨
1  2017-12-30 00:00   서울 강남구  서울  강남구  1168000000   23.29       나쁨
2  2017-12-29 00:00   서울 강남구  서울  강남구  1168000000   51.60       보통
3  2017-12-28 00:00   서울 강남구  서울  강남구  1168000000   41.26       보통
4  2017-12-27 00:00   서울 강남구  서울  강남구  1168000000   39.92       나쁨

## ~~한국관광공사_관광빅데이터(우선 주석처리)~~

In [ ]:
# SV3 = join(DATA_PATH, "한국관광공사_관광빅데이터")
# %cd "{SV3}"
# !ls

/content/drive/.shortcut-targets-by-id/1TVtWvZ4DNn7V20Aj6JzSd5_iRdbjEXbG/Proj_WT/DataSets/한국관광공사_관광빅데이터
시도별_방문자수집계_0426_0618.csv    서울자치구_방문자수집계_0426_0627.csv
시군구별_방문자수집계_0426_0627.csv


In [ ]:
# ref = join(DATA_PATH, "api활용_참고_문서", "한국관광공사_관광_빅데이터_정보_서비스")
# %cd "{ref}"
# !ls

# serviceCD = pd.read_excel(f"{ref}/한국관광공사_서비스코드(관광빅데이터)_v1.0.xlsx")

# seoul = serviceCD.iloc[:25]
# seoul

/content/drive/.shortcut-targets-by-id/1TVtWvZ4DNn7V20Aj6JzSd5_iRdbjEXbG/Proj_WT/DataSets/api활용_참고_문서/한국관광공사_관광_빅데이터_정보_서비스
'한국관광공사_TourAPI활용신청방법_매뉴얼_v3.2(20180323).docx'
'한국관광공사_활용매뉴얼(관광빅데이터)_v1.0.docx'
'한국관광공사_서비스코드(관광빅데이터)_v1.0.xlsx'


areaCode areaNm  signguCode signguNm
0         11  서울특별시       11110      종로구
1         11  서울특별시       11140       중구
2         11  서울특별시       11170      용산구
3         11  서울특별시       11200      성동구
4         11  서울특별시       11215      광진구
5         11  서울특별시       11230     동대문구
6         11  서울특별시       11260      중랑구
7         11  서울특별시       11290      성북구
8         11  서울특별시       11305      강북구
9         11  서울특별시       11320      도봉구
10        11  서울특별시       11350      노원구
11        11  서울특별시       11380      은평구
12        11  서울특별시       11410     서대문구
13        11  서울특별시       11440      마포구
14        11  서울특별시       11470      양천구
15        11  서울특별시       11500      강서구
16        11  서울특별시       11530      구로구
17        11  서울특별시       11545      금천구
18        11  서울특별시       11560     영등포구
19        11  서울특별시       11590      동작구
20        11  서울특별시       11620      관악구
21        11  서울특별시       11650      서초구
22        11  서울특별시       11680      강남구
23        11  서울특별시       11710      송파구
24        11  서울특별시       11740      강동구

#### 서울 자치구 코드 및 자치구 명 불러오기

In [ ]:
# seoul_sggCD = [code for code in seoul.iloc[:, -2]]
# seoul_sggNM = [name for name in seoul.iloc[:, -1]]

### 전체 방문자수집계 분석 시 한 번에 파일 불러오기(필요 시 주석 제거)

In [ ]:
# #광역지자체
# metro = "metcoRegnVisitrDDList"

# local_df = pd.read_csv(f'{DATA_PATH}/한국관광공사_관광빅데이터/시도별_방문자수집계_0426_0618.csv')

In [ ]:
# #기초지자체
# local = "locgoRegnVisitrDDList"

# local_df = pd.read_csv(f'{DATA_PATH}/한국관광공사_관광빅데이터/시군구별_방문자수집계_0426_0618.csv')

### 서울 관광지방문자수집계 자료만 파일 불러오기

In [ ]:
# local_seoul = pd.read_csv(f'{SV3}/서울자치구_방문자수집계_0426_0618.csv')

# EDA & VSL

## EDA(Match and Merge)

### 추가 라이브러리 및 함수

In [ ]:
#when dtype is int or float
def zero_Check(dt, col):
  zrcnt = dt[col][df[col]==0].count()
  zridx = dt[df[col]==0].index
  print(f'zero counts of {col} : {zrcnt}\n')
  print(f'\t{zridx}')
  return zrcnt, zridx

#when dtype is object
def null_Check(dt, col):
  zrcnt = 0
  zridx = []
  for idx, cont in enumerate(dt[col]):
    if type(cont)==float:
      zrcnt += 1
      zridx.append(idx)
  print(f'null values in {col} : {zrcnt}\n')
  print(f'\t{zridx}')
  return zrcnt, zridx

#no matter dtype numeric nor category, check value is string typed "N/A"
def NONE_Check(dt, col):
  nacnt = dt[col][df[col]==("None" or "N" or "NONE" or "N/A")].count()
  naidx = dt[df[col]==("None" or "N" or "NONE" or "N/A")].index
  print(f'"N/A" counts of {col} : {nacnt}\n')
  print(f'\t{naidx}')
  return nacnt, naidx



### 상기 과정에서 간단히 정제한 데이터 확인

In [ ]:
#울산 기상 데이터(11~19년)
print(Ulsan_WT.shape)
print(display(Ulsan_WT.head(2)))

#울산 주요관광지점 입장객 데이터
  ##연별 데이터
print(f"울산 주요관광지점 입장객 연별 데이터 : \n\t\t\t\t\t shape of data : {YR_Ulsan.shape}\n")
print(display(YR_Ulsan.head(2)))
  ##연월별 데이터
print(f"울산 주요관광지점 입장객 연월별 데이터 : \n\t\t\t\t\t shape of data : {YM_Ulsan.shape}\n")
print(display(YM_Ulsan.head(2)))

(2957, 62)


지점 번호 지점명         일시  평균 기온(°C)  ...  합계 소형증발량(mm)  9-9강수(mm)  일기현상  안개 계속 시간(hr)
0    152  울산 2011-01-01       -1.4  ...           NaN        NaN   NaN           NaN
1    152  울산 2011-01-02        1.8  ...           NaN        NaN   NaN           NaN

[2 rows x 62 columns]

None
울산 주요관광지점 입장객 연별 데이터 : 
					 shape of data : (9, 30)



2      3    4    5    6   ...      52     56      57     58     61
2011년  74971  32227  NaN  NaN  NaN  ...  359853  46927  102915    NaN  49243
2012년  82505  42306  NaN  NaN  NaN  ...  364107  38492   92190  14869  47631

[2 rows x 30 columns]

None
울산 주요관광지점 입장객 연월별 데이터 : 
					 shape of data : (108, 24)



2     3    4    5    6     8   ...    45   47     52    56    57   58
2011년 01월  6107  2727  NaN  NaN  NaN  5198  ...  2350  NaN  50378  2960  6537  NaN
2011년 02월  5920  1696  NaN  NaN  NaN  3589  ...  2873  NaN  24037  3852  6210  NaN

[2 rows x 24 columns]

None


### 기상데이터 일>월 결측치 확인(변수명 : Ulsan_WT<small>(updated)</small>)

* 목적 : 주요관광지점 입장객 수와 붙이기

In [ ]:
365*9

3285

In [ ]:
len(Ulsan_WT["일시"])

2957

In [ ]:
Ulsan_WT[Ulsan_WT.loc[:,"일시"]<="2011-12-31"]

# DataFrame.at : Access a single value for a row/column label pair.
# DataFrame.iloc : Access group of rows and columns by integer position(s).
# DataFrame.xs : Returns a cross-section (row(s) or column(s)) from the
#     Series/DataFrame.
# Series.loc : Access group of values using labels.

지점 번호 지점명  ...                                               일기현상  안개 계속 시간(hr)
0      152  울산  ...                                                NaN           NaN
1      152  울산  ...                                                NaN           NaN
2      152  울산  ...  {눈}1110-{눈}{강도0}1200-{눈}{강도1}1500-{눈}{강도0}1800...           NaN
3      152  울산  ...  -{진눈깨비}-{눈}0010-0040. -{박무}-{박무}{강도0}0300-{박무}...           NaN
4      152  울산  ...                                                NaN           NaN
..     ...  ..  ...                                                ...           ...
360    152  울산  ...                                                NaN           NaN
361    152  울산  ...                                                NaN           NaN
362    152  울산  ...                                                NaN           NaN
363    152  울산  ...                                    {무지개}1510-1610.           NaN
364    152  울산  ...                       {박무}0520-{박무}{강도0}0600-0840.           NaN

[365 rows x 62 columns]

단순히 365일을 기준으로 9년을 곱했을 때, 3285이란 결과가 나옴.

현재 기상데이터는 2957에 대한 정보만 존재하므로, 결측일(혹은 결측월)에 대한 정보를 알고 있어야 <b><big>해석의 용의성</big> 및 <big>모델의 정확도</big></b>가 올라감

윤년이란 상황을 가정하고, 매월 달라지는 2월의 일자까지 확인하기 위해서 우선 datetime 모듈을 활용해서 일 계산 및 결측치 확인에 사용

In [ ]:
def Date_Duration():
  print(f"{'='*10}{'<'*5}\tInitializing\t{'>'*5}{'='*300}")
  sY, sM, sD  = map(int, input("시작일 : ").split(" "))
  startDate = datetime(sY, sM, sD)

  eY, eM, eD  = map(int, input("종료일 : ").split(" "))
  endDate = datetime(eY, eM, eD)

  duration = (endDate-startDate).days
  print(f"조사 기간의 기상 데이터의 일 수는 총 {duration}일이어야 합니다.\n")
  #결측일이 없으려면, duration 값과 기상 데이터의 rows가 동일해야 함

  return startDate, endDate, duration


def Missing_Date_Check(dt, datecol, duration, startDate=None, check_type="%Y%m%d"):
  print(f"{'='*10}{'<'*5}\tmissing dates\t{'>'*5}{'='*300}")
  missing_days = []
  missing_day = 0

  for day in range(duration):
    check_date = startDate+timedelta(days=day)
    if dt[datecol][day-missing_day].strftime(check_type)==check_date.strftime(check_type):
      continue
    elif dt[datecol][day-missing_day].strftime(check_type)!=check_date.strftime(check_type):
      missing_day+=1
      missing_days.append(check_date) # 활용을 쉽게 활용할 수 있도록 datetime으로 추가

  if len(missing_days)!=0:
    print(f"*경고*\n결측일이 발견되었습니다.")
    print(f"{[mdate.strftime(check_type) for mdate in missing_days]}의 결측일 여부를 확인하세요.\n")

  return missing_days


def Re_val_MDate(dt, datecol, missing_days, check_type="%Y%m%d"):
  print(f"{'='*10}{'<'*5}\tre-validation\t{'>'*5}{'='*300}")
  cnt = 0
  re_val_date = []

  for idx, mdate in enumerate(missing_days):
    if len(dt[dt[datecol]==mdate.strftime(check_type)].index)!=0:
      re_val_date.append(idx)
    elif len(dt[dt[datecol]==mdate.strftime(check_type)].index)==0:
      cnt += 1

  if cnt==len(missing_days):
    print("재검증 결과 : 상동")
    print(f"해당 기상 데이터에 존재하는 결측일은 총 {cnt}일입니다.\n")
  elif cnt!=len(missing_days):
    for pop_idx in re_val_date:
      missing_days.pop(pop_idx)
    print("재검증 결과 : 상이")
    print(f"해당 기상 데이터에 존재하는 결측일은 총 {len(missing_days)}일입니다.\n")

  return missing_days


def Cnts_MDays_in_YearMonth(missing_days, check_ym_type="%Y%m", return_ym_type="%Y년%m월"):
  print(f"{'='*10}{'<'*5}\tCounts of Missing days by Months\t{'>'*5}{'='*300}")

  if check_ym_type!=return_ym_type:
    missing_days_by_ym = [mdates.strftime(return_ym_type) for mdates in missing_days]
    missing_ym = sorted(list(set(missing_days_by_ym)))
    for mYM in missing_ym:
      globals()[f"missing_{mYM}"] = missing_days_by_ym.count(mYM)
      print(f'{mYM}의 결측일 수 : {globals()[f"missing_{mYM}"]}\n')

  else :
    missing_days_by_ym = [mdates.strftime(check_ym_type) for mdates in missing_days]
    missing_ym = sorted(list(set(missing_days_by_ym)))
    for mYM in missing_ym:
      globals()[f"missing_{mYM}"] = missing_days_by_ym.count(mYM)
      print(f'{mYM}의 결측일 수 : {globals()[f"missing_{mYM}"]}\n')

  return missing_ym

In [ ]:
# 관측기간 시작일 및 종료일 입력 후 rows 맞는지 확인
stD, edD, during = Date_Duration()
# 기상 데이터 칼럼명으로 접근해서 결측일 조회 및 검증
odd_days = Missing_Date_Check(Ulsan_WT, "일시", during, stD)
odd_days = Re_val_MDate(Ulsan_WT, "일시", odd_days, "%Y-%m-%d")
# 주요관광지점 입장객 수 데이터의 "%Y년%m월"과 동일한 형태
  # 기상데이터 월별 index 혹은 columns으로 활용
odd_ym = Cnts_MDays_in_YearMonth(odd_days)

==========<<<<<	Initializing	>>>>>============================================================================================================================================================================================================================================================================================================
시작일 : 2011 1 1
종료일 : 2019 12 31
조사 기간의 기상 데이터의 일 수는 총 3286일이어야 합니다.

==========<<<<<	missing dates	>>>>>============================================================================================================================================================================================================================================================================================================
*경고*
결측일이 발견되었습니다.
['20170411', '20170412', '20170413', '20170414', '20170415', '20170416', '20170417', '20170418', '20170419', '20170420', '20170421', '20170422', '20170423', '20170424', '20170425', '20170426', '20170427', '20170428', '20170429', '20170430', 


* 궁금하면 확인해보기 1

```
# print((Ulsan_WT["일시"][0]).strftime('%Y%m%d'), type((Ulsan_WT["일시"][0]).strftime('%Y%m%d')))
# print((startDate+timedelta(days=0)).strftime('%Y%m%d'), type((startDate+timedelta(days=0)).strftime('%Y%m%d')))
```

* 궁금하면 확인해보기 2

```
# type(Ulsan_WT["일시"][day])==Timestamp
# type(startDate+timedelta(days=day))==datetime.datetime
```

* 궁금하면 확인해보기 3

```
# 결측일 발견 여부 재검증
cnt = 0
errored_date = []

for idx, missing_date in enumerate(odd_days):
  if len(Ulsan_WT[Ulsan_WT["일시"]==missing_date.strftime('%Y-%m-%d')].index)!=0:
    print(f"{missing_date}의 결측일 여부는 재확인이 필요합니다.\n\n")
    errored_date.append(idx)
  elif len(Ulsan_WT[Ulsan_WT["일시"]==missing_date.strftime('%Y-%m-%d')].index)==0:
    cnt += 1

if cnt==len(odd_days):
  print("변수 odd_days에 저장된 일자는 해당 기상 데이터에서는 불러올 수 없는 결측치입니다.")
  print(f"해당 기상 데이터에 존재하는 결측일은 총 {cnt}일입니다.\n\n")
elif cnt!=len(odd_days):
  for pop_mistakes in errored_date:
    odd_days.pop(cnt)
  print("변수 odd_days에 저장된 일자는 해당 기상 데이터에서는 불러올 수 없는 결측치입니다.")
  print(f"해당 기상 데이터에 존재하는 결측일은 총 {cnt}일입니다.\n\n")
```

* 궁금하면 확인해보기 4

```
missing_ym = sorted(list(set([ym.strftime("%Y년%m월") for ym in odd_days])))

# missing_ym.index("201704") ## 0 ### cf. 리스트도 index로 접근 가능

for lst in missing_ym:
  locals()[f"mYM_{lst}"] = []

for mdate in odd_days:
  check_mdate = mdate.strftime("%Y%m%d")
  chk_md = check_mdate[:4]+"년"+check_mdate[4:6]+"월"

  if chk_md in missing_ym:
    locals()[f"mYM_{chk_md}"].append(check_mdate)

for lst in missing_ym:
  print(f"\t\t\t[{lst}]의 결측일 수는 총 {len(locals()[f'mYM_{lst}'])}입니다.")

# 해당 월의 데이터가 전부 없다면, 해당 월은 관광지와 merge 시키기 전에 관광지 데이터에서도 월 데이터를 지우고 계산.
# 해당 월의 데이터가 소실된 양이 많다면, 정확도를 위해서는 그 데이터를 뜯어야 겠지만...우선은 mean으로 연산  
```

* 궁금하면 확인해보기 5

```
odd_ym = sorted(list(set([ym.strftime("%Y년%m월") for ym in odd_days])))
check_lst_ym = [mdates.strftime("%Y년%m월") for mdates in odd_days]
print(check_lst_ym)
print(odd_ym)
for cnts in odd_ym:
  print(check_lst_ym.count(cnts))
```

* 해당 월의 데이터가 전부 없다면, 해당 월은 관광지와 merge 시키기 전에 관광지 데이터에서도 월 데이터를 지우고 계산.
* 해당 월의 데이터가 소실된 양이 많다면, 정확도를 위해서는 그 데이터를 뜯어야 겠지만...우선은 mean으로 연산 

In [ ]:
# 2018년 12월 데이터가 있어야 하는 열
Ulsan_WT[-365-31:-365]

울산 기상데이터의 결측일을 조회한 상기의 결과로 봤을 때,

* 2017년
  * 05월
  * 06월

* 2018년
  * 03월
  * 04월
  * 05월
  * 06월
  * 07월
  * 12월

은 제외하고 울산 주요관광지점 입장객 수와 합쳐야 함

데이터 확인을 먼저 해야할 지,,, 아니면 15일 이상 결측치가 있어보이는데,,, 다 날릴까???

datetime모듈을 사용해 함수로 구현하면 빠른 접근 가능(나중 문제)

### 기상데이터 일>월 월별데이터 정제

In [ ]:
def Sep_DF_by_Year(dt, col, count)

In [ ]:
# def Cnts_Days_in_YearMonth(date_Series, check_ym_type="%Y%m", return_ym_type="%Y년%m월"):
#   print(f"{'='*10}{'<'*5}\tCounts of days by Months\t{'>'*5}{'='*300}")
#   # date_Series

#   if check_ym_type!=return_ym_type:
#     days_by_ym = [mdates.strftime(return_ym_type) for mdates in missing_days]
#     ym = sorted(list(set(missing_days_by_ym)))
#     for mYM in missing_ym:
#       globals()[f"days_{mYM}"] = missing_days_by_ym.count(mYM)
#       print(f'{mYM}의 결측일 수 : {globals()[f"missing_{mYM}"]}\n')

#   else :
#     missing_days_by_ym = [mdates.strftime(check_ym_type) for mdates in missing_days]
#     missing_ym = sorted(list(set(missing_days_by_ym)))
#     for mYM in missing_ym:
#       globals()[f"missing_{mYM}"] = missing_days_by_ym.count(mYM)
#       print(f'{mYM}의 결측일 수 : {globals()[f"missing_{mYM}"]}\n')

#   return missing_ym

### 주요관광지점 입장객 수 데이터>기상데이터 결측치 데이터 drop(axis=0)

In [ ]:
Check_info.iloc[[1,2] ,:]

0            1           2   ...   59       60         61
군구           중구           중구          중구  ...  울주군      울주군        울주군
관광지  수운 최제우 유허지  약사동제방유적전시관   외솔생가 및 기념관  ...  작천정  진하 해수욕장  충렬공박제상기념관

[2 rows x 62 columns]

In [ ]:
YM_UorgCol = [col for col in YM_Ulsan.columns]
YM_UnewCol = [area+"_"+sgg for sgg, area in zip(Check_info.iloc[1 ,:], Check_info.iloc[2 ,:])]

YM_UfinCol = {}
for idx in range(len(YM_Ulsan.columns)):
  YM_UfinCol[YM_UorgCol[idx]]=YM_UnewCol[idx]

YM_Ulsan.rename(columns=YM_UfinCol, inplace=True)

YM_Ulsan.reset_index(inplace=True)

YM_Ulsan.rename(columns={"index":"일시"}, inplace=True)

/usr/local/lib/python3.7/dist-packages/pandas/core/frame.py:4308: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [ ]:
YM_Ulsan

일시 수운 최제우 유허지_중구 약사동제방유적전시관 _중구  ... 대왕암공원_동구 소리체험관_동구 슬도_동구
0    2011년 01월          6107           2727  ...     2960     6537   NaN
1    2011년 02월          5920           1696  ...     3852     6210   NaN
2    2011년 03월          6617           2924  ...     3970     7081   NaN
3    2011년 04월          7115           3339  ...     4922    11598   NaN
4    2011년 05월          7173           3472  ...     6361    11165   NaN
..         ...           ...            ...  ...      ...      ...   ...
103  2019년 08월          3194           3835  ...     4618    10059  1863
104  2019년 09월          4208           4133  ...     5674     7005  1241
105  2019년 10월         11857           7722  ...     4902    10255  1908
106  2019년 11월          3685           4358  ...     4852     9387  1239
107  2019년 12월          2392           3429  ...     2360     7178  1339

[108 rows x 25 columns]

## EDA(Statistics)

* heatmap
* statsmodel's ols

## EDA(Tourism Climate)

### Tourism Climate

* De Freitas(2003)
관광기후학 측면에서 이론적인 개념 시도
  * 열적(Thermal)측면

    관광자의 쾌적함(편안하게 느끼는)과 기온, 바람, 습도 등이 복합적으로 결합된 것

  * 물리적(Physical)측면

    바람이나 강수처럼 관광활동에 직접적으로
  제약요인으로 작용하는 것

  * 미적(Aesthetic)측면

    관광 자의 감정이나 경험의 질에 영향을 주는 풍경과 연관된 것으로서 파란하늘, 구름 의 양, 햇빛, 시야, 일조시간 등을 의미

세 가지 요소는 동시에 작용되기 때문에 날씨, 기후의 최적조건에 관한 연구는 세 요소를 고려해야 함

* Gomez-martin(2005)

  상기의 요소(물리적(비, 바람 등), 생리적(기온, 습도 등), 심리적(구름의 양, 일조 시간 등)) + 지리학적 관점

  관광활동 참가자 및 관광지 방문객의
  불만족을 줄이기 위해 수요자 중심의 다양한 날씨, 기후 정보를 신속하게
  제공해야 함을 지적

이들의 논의는 앞선 두 연구자들의 견해,즉 날씨, 기후는 관 광자의 활동에 영향을 미치고 만족도를 높이기 위해서 이를 관리할
필요가 있다는 견해를 확장한 것으로 볼 수 있다. 따라서 관광활동 및 관광지 특성에 적합한 날씨, 기후 정보가 만들어지고 관광자에게 제공될 필요가 있다.

* TCI(Mieczkowski_1985년)

  도시, 야외에서 걷기, 관람, 쇼핑 등의 일반 관광과 관련된 지수

  단, 해변, 스키 등의 활동에는 적용하지 못하는 한계를 지님

  * TCI
    * 열적 쾌적성
      * Cd : 평균기온, 평균상대습도
      * Ca : 최고기온, 최저상대습도
    * 강수(P) : 총강수량(mm)
    * 풍속(W) : 평균풍속(m/s)
    * 일조시간(S) : 일조시간(hr)

  총 7개의 기후 변수로 관광에 적합한 날인지 아닌지를 파악하는 지표

* 한국형 관광기후지수(KTCI)

  4계절을 통합한 기후 가중치의 경우 최고기온은 25.9%, 평균기온은 20.9%, 강수량은 32.2%, 풍속은 11.6%, 구름의 양은 9.5%로 나타났다. 한국의 경우 강수량이 가장 높은 가중치를 가진 기후요소인 반면에 구름의 양은 가장 낮은 가중치로 나타났다. 또한 본 연구에서는 계절별로 또한 야외여가활동별로 기후요소별 가중치의 차이를 이용한 더욱 세부적인 한국형 TCI를 제안하였다. 본 연구에서 도출된 KTCI는 우리나라의 관광자과 관광사업체에게 관광 및 여가활동을 하기에 좋은(또는 나쁜) 시기와 공간적 분포를 분석하는데 기본적인 지표로 활용될 수 있다. 또한 기후예보나 기후변화 예측을 통해 관광객의 여가활동에 대한 예측과 관광공간 변화 등을 예측하는 분야에서도 사용될 수 있을 것으로 본다.

#### 요섭님께서 작성하신 함수

In [ ]:
class Input(self):
  def tour_index(x1, x2, y1, y2):
    e = ((y2-y1)/(y2+y1))/((x2-x1)/(x2+x1))
    e = pd.DataFrame(e, columns=["편차"])
    return e

  def output_index(y1, y2):
    a1 = self["총 평균 평균기온(°C)"]
    a2 = self["평균기온(°C)"]
    a = Input.tour_index(a1, a2, y1, y2)

    b1 = self["총 평균 최고기온(°C)"]
    b2 = self["최고기온(°C)"]
    b = Input.tour_index(b1, b2, y1, y2)

    c1 = self["총 평균 평균상대습도(%)"]
    c2 = self["평균상대습도(%)"]
    c = Input.tour_index(c1, c2, y1, y2)

    d1 = self["총 평균 최저상대습도(%)"]
    d2 = self["최저상대습도(%)"]
    d = Input.tour_index(d1, d2, y1, y2)

    e1 = self["총 평균 일강수량(mm)"]
    e2 = self["일강수량(mm)"]
    e = Input.tour_index(e1, e2, y1, y2)

    f1 = self["총 평균 평균풍속(m/s)"]
    f2 = self["평균풍속(m/s)"]
    f = Input.tour_index(f1, f2, y1, y2)

    g1 = self["총 평균 합계 일조 시간(hr)"]
    g2 = self["합계 일조 시간(hr)"]
    g = Input.tour_index(g1, g2, y1, y2)

    TCI = (a+b+c+d+e+f+g)/7
    return TCI




#### 요섭님께서 작성하신 함수(수정 중)

In [ ]:
class cal_TourIDX:
  def __init__(self, chk_col):
    data = self
    chk_col = dict()
    cnt = 0
  
  def check_columns(y1, y2):
    cnt += 1
    chk_col[cnt] = list(y1, y2)

  def tour_index(x1, x2, y1, y2):
    e = ((y2-y1)/(y2+y1))/((x2-x1)/(x2+x1))
    e = pd.DataFrame(e, columns=["편차"])
    return e

KTCI_col = ["총 평균 평균기온(°C)", "평균기온(°C)"]

  def output_index(y1, y2):
    a1 = self["총 평균 평균기온(°C)"]
    a2 = self["평균기온(°C)"]
    a = Input.tour_index(a1, a2, y1, y2)

    b1 = self["총 평균 최고기온(°C)"]
    b2 = self["최고기온(°C)"]
    b = Input.tour_index(b1, b2, y1, y2)

    c1 = self["총 평균 평균상대습도(%)"]
    c2 = self["평균상대습도(%)"]
    c = Input.tour_index(c1, c2, y1, y2)

    d1 = self["총 평균 최저상대습도(%)"]
    d2 = self["최저상대습도(%)"]
    d = Input.tour_index(d1, d2, y1, y2)

    e1 = self["총 평균 일강수량(mm)"]
    e2 = self["일강수량(mm)"]
    e = Input.tour_index(e1, e2, y1, y2)

    f1 = self["총 평균 평균풍속(m/s)"]
    f2 = self["평균풍속(m/s)"]
    f = Input.tour_index(f1, f2, y1, y2)

    g1 = self["총 평균 합계 일조 시간(hr)"]
    g2 = self["합계 일조 시간(hr)"]
    g = Input.tour_index(g1, g2, y1, y2)

    TCI = (a+b+c+d+e+f+g)/7
    return TCI


### 기상 데이터 분류

#### 열적 측면 기상 데이터(변수명 : UWT_Ther)

생리학적 측면 기상 데이터라고도 불리지만, 열적 측면 기상데이터라고 명명하는 것이 이해를 돕기 때문에 열적 측명 기상데이터라고 명명함

In [ ]:
Ulsan_WT

In [ ]:
#열적 측면 기상 데이터
UWT_Ther

#### 물리적 측면 기상 데이터(변수명 : UWT_Phys)


In [ ]:
#물리적 측면 기상 데이터
UWT_Phys

#### 미적 측면 기상 데이터(변수명 : UWT_Aest)

심리적 측면 기상데이터라고도 불리나, 미적 측면 기상 데이터로 명명함이 관광적인 측면에서는 이해가 높아짐

In [ ]:
#미적 측면 기상 데이터
UWT_Aest

#### 지리학적 측면 기상 데이터(변수명 : UWT_Geog)

In [ ]:
#지리학적 측면 기상 데이터
UWT_Geog

#### TCI 기상 데이터(변수명 : UWT_TCI)

In [ ]:
#TCI 기상 데이터
UWT_TCI

#### KTCI 기상 데이터(변수명 : UWT_KTCI)

In [ ]:
#KTCI 기상 데이터
UWT_KTCI

날씨(기후)를 가지고만 추천을 하려고 하는데, 상기의 방법으로 풀었을 때, 결과가 안 나온다면 어떻게 해야 하는가?

## VSL

### 라이브러리 및 함수 작성

In [ ]:
#정열님 sns 함수

def graph(data = None, x = None, y = None, hue = None, option = None):
  data = data[data[y] != np.nan]
  data[y] = data[y].astype('float64')
  if option == 'line':
    plt.figure(figsize = (16,8))
    sns.lineplot(data = data, x = x, y = y, hue = hue)
  elif option == 'bar':
    plt.figure(figsize = (16,8))
    sns.barplot(data = data, x = x, y = y, hue = hue)
  elif option == 'box':
    plt.figure(figsize = (16,8))
    sns.boxenplot(data = data, x = x, y = y, hue = hue)
  elif option == 'dist':
    plt.figure(figsize = (16,8))
    sns.distplot(data = data, x = x, y = y, hue = hue)
  elif option == 'count':
    plt.figure(figsize = (16,8))
    sns.countplot(data = data, x = x, y= y, hue = hue)
  else:
    pass
  plt.title(f'{y}')
  plt.show()

# Model

## Machine Learning

### Baseline

In [ ]:
# 라이브러리 및 함수 작성


### PCA(Principal Componets Analysis)

주성분 분석으로 고차원 데이터를 저차원 데이터로 축소시키는 방법

### SVD(Singular Value Decomposition)

특이값 분해를 사용해 저차원 데이터로 축소시키는 방법(신호 처리와 통계학 등의 분야에서 자주 사용)

## Deap Learning

### Baseline

In [ ]:
# 라이브러리 및 함수 작성
import statsmodels.api as sm

### VAR(Vector Auto Regression)

예측할 변수의 과거 값 뿐만 아니라 예측할 변수의 의존성이 있는 변수들까지 고려하여 선형 함수로 나타내는 확률적 과정

In [ ]:
from statsmodels.tsa.api import VAR
from statsmodels.tsa.stattools import adfuller

# copy files__다른 모델, 추가 가설 증명 시 활용

In [ ]:
Basic = PROJ_PATH
fexe = ".ipynb"
origin = "울산_날씨_관광_분석"+fexe
org = join(Basic, origin)

copy1 = "강원_날씨_관광_분석"+fexe
cps1 = join(Basic, copy1)

copy2 = "경기_날씨_관광_분석"+fexe
cps2 = join(Basic, copy2)

copy3 = "광주_날씨_관광_분석"+fexe
cps3 = join(Basic, copy3)

copy4 = "날씨_관광_분석_담당전체"+fexe
cps4 = join(Basic, copy4)
!cp "{org}" "{cps}"
# 디렉토리에서 파일을 옮기는 경우, 디렉토리와 파일명이 같이 입력되어야 함
# 공유 폴더 내 존재하는 디렉토리 명은 상기에 다 입력했음(2021.06.25 기준)

In [ ]:
""